// n8n-nodes-1shot/dist/nodes/OneShot/crypto/ED25519.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyAsync = exports.verify = exports.utils = exports.signAsync = exports.sign = exports.Point = exports.getPublicKeyAsync = exports.getPublicKey = exports.ExtendedPoint = exports.etc = exports.CURVE = void 0;
const crypto_1 = require("crypto");
const P = 2n ** 255n - 19n;
const N = 2n ** 252n + 27742317777372353535851937790883648493n;
const Gx = 0x216936d3cd6e53fec0a4e231fdd6dc5c692cc7609525a7b2c9562d608f25d51an;
const Gy = 0x6666666666666666666666666666666666666666666666666666666666666658n;
const _a = P - 1n;
const _d = 0x52036cee2b6ffe738cc740797779e89800700a4d4141d8ab75eb4dca135978a3n;
const h = 8n;
const MASK = 2n ** 256n;
const L = 32;
const L2 = L * 2;
const CURVE = {
    a: _a,
    d: _d,
    p: P,
    n: N,
    h: h,
    Gx: Gx,
    Gy: Gy,
};
exports.CURVE = CURVE;
const err = (m = '') => {
    throw new Error(m);
};
const isS = (s) => typeof s === 'string';
const isB = (s) => typeof s === 'bigint';
const isu8 = (a) => a instanceof Uint8Array || (ArrayBuffer.isView(a) && a.constructor.name === 'Uint8Array');
const abytes = (a, l) => !isu8(a) || (typeof l === 'number' && l > 0 && a.length !== l)
    ? err('Uint8Array of valid length expected')
    : a;
const u8n = (len) => new Uint8Array(len);
const u8fr = (buf) => Uint8Array.from(buf);
const toU8 = (a, len) => abytes(isS(a) ? hexToBytes(a) : u8fr(abytes(a)), len);
const M = (a, b = P) => {
    let r = a % b;
    return r >= 0n ? r : b + r;
};
const arange = (n, min, max = MASK, msg = 'bad number: out of range') => (isB(n) && min <= n && n < max ? n : err(msg));
const apoint = (p) => (p instanceof Point ? p : err('Point expected'));
class Point {
    constructor(ex, ey, ez, et) {
        this.ex = arange(ex, 0n);
        this.ey = arange(ey, 0n);
        this.ez = arange(ez, 1n);
        this.et = arange(et, 0n);
        Object.freeze(this);
    }
    static fromAffine(p) {
        return new Point(p.x, p.y, 1n, M(p.x * p.y));
    }
    static fromBytes(hex, zip215 = false) {
        const d = _d;
        abytes(hex, L);
        const normed = hex.slice();
        const lastByte = hex[31];
        normed[31] = lastByte & ~0x80;
        const y = bytesToNum_LE(normed);
        const max = zip215 ? MASK : P;
        arange(y, 0n, max);
        const y2 = M(y * y);
        const u = M(y2 - 1n);
        const v = M(d * y2 + 1n);
        let { isValid, value: x } = uvRatio(u, v);
        if (!isValid)
            err('bad y coord 3');
        const isXOdd = (x & 1n) === 1n;
        const isLastByteOdd = (lastByte & 0x80) !== 0;
        if (!zip215 && x === 0n && isLastByteOdd)
            err('bad y coord 4');
        if (isLastByteOdd !== isXOdd)
            x = M(-x);
        return new Point(x, y, 1n, M(x * y));
    }
    static fromHex(hex, zip215) {
        return Point.fromBytes(toU8(hex), zip215);
    }
    get x() {
        return this.toAffine().x;
    }
    get y() {
        return this.toAffine().y;
    }
    assertValidity() {
        const a = _a;
        const d = _d;
        const p = this;
        if (p.is0())
            throw new Error('bad point: ZERO');
        const { ex: X, ey: Y, ez: Z, et: T } = p;
        const X2 = M(X * X);
        const Y2 = M(Y * Y);
        const Z2 = M(Z * Z);
        const Z4 = M(Z2 * Z2);
        const aX2 = M(X2 * a);
        const left = M(Z2 * M(aX2 + Y2));
        const right = M(Z4 + M(d * M(X2 * Y2)));
        if (left !== right)
            throw new Error('bad point: equation left != right (1)');
        const XY = M(X * Y);
        const ZT = M(Z * T);
        if (XY !== ZT)
            throw new Error('bad point: equation left != right (2)');
        return true;
    }
    equals(other) {
        const { ex: X1, ey: Y1, ez: Z1 } = this;
        const { ex: X2, ey: Y2, ez: Z2 } = apoint(other);
        const X1Z2 = M(X1 * Z2), X2Z1 = M(X2 * Z1);
        const Y1Z2 = M(Y1 * Z2), Y2Z1 = M(Y2 * Z1);
        return X1Z2 === X2Z1 && Y1Z2 === Y2Z1;
    }
    is0() {
        return this.equals(I);
    }
    negate() {
        return new Point(M(-this.ex), this.ey, this.ez, M(-this.et));
    }
    double() {
        const { ex: X1, ey: Y1, ez: Z1 } = this;
        const a = _a;
        const A = M(X1 * X1);
        const B = M(Y1 * Y1);
        const C = M(2n * M(Z1 * Z1));
        const D = M(a * A);
        const x1y1 = X1 + Y1;
        const E = M(M(x1y1 * x1y1) - A - B);
        const G = D + B;
        const F = G - C;
        const H = D - B;
        const X3 = M(E * F);
        const Y3 = M(G * H);
        const T3 = M(E * H);
        const Z3 = M(F * G);
        return new Point(X3, Y3, Z3, T3);
    }
    add(other) {
        const { ex: X1, ey: Y1, ez: Z1, et: T1 } = this;
        const { ex: X2, ey: Y2, ez: Z2, et: T2 } = apoint(other);
        const a = _a;
        const d = _d;
        const A = M(X1 * X2);
        const B = M(Y1 * Y2);
        const C = M(T1 * d * T2);
        const D = M(Z1 * Z2);
        const E = M((X1 + Y1) * (X2 + Y2) - A - B);
        const F = M(D - C);
        const G = M(D + C);
        const H = M(B - a * A);
        const X3 = M(E * F);
        const Y3 = M(G * H);
        const T3 = M(E * H);
        const Z3 = M(F * G);
        return new Point(X3, Y3, Z3, T3);
    }
    mul(n, safe = true) {
        if (n === 0n)
            return safe === true ? err('cannot multiply by 0') : I;
        arange(n, 1n, N);
        if (n === 1n || (!safe && this.is0()))
            return this;
        if (this.equals(G))
            return wNAF(n).p;
        let p = I, f = G;
        for (let d = this; n > 0n; d = d.double(), n >>= 1n) {
            if (n & 1n)
                p = p.add(d);
            else if (safe)
                f = f.add(d);
        }
        return p;
    }
    multiply(scalar) {
        return this.mul(scalar);
    }
    clearCofactor() {
        return this.mul(BigInt(h), false);
    }
    isSmallOrder() {
        return this.clearCofactor().is0();
    }
    isTorsionFree() {
        let p = this.mul(N / 2n, false).double();
        if (N % 2n)
            p = p.add(this);
        return p.is0();
    }
    toAffine() {
        const { ex: x, ey: y, ez: z } = this;
        if (this.equals(I))
            return { x: 0n, y: 1n };
        const iz = invert(z, P);
        if (M(z * iz) !== 1n)
            err('invalid inverse');
        return { x: M(x * iz), y: M(y * iz) };
    }
    toBytes() {
        const { x, y } = this.toAffine();
        this.assertValidity();
        const b = numToBytes_32LE(y);
        b[31] |= x & 1n ? 0x80 : 0;
        return b;
    }
    toHex() {
        return bytesToHex(this.toBytes());
    }
    toRawBytes() {
        return this.toBytes();
    }
}
exports.ExtendedPoint = Point;
exports.Point = Point;
const G = new Point(Gx, Gy, 1n, M(Gx * Gy));
const I = new Point(0n, 1n, 1n, 0n);
Point.BASE = G;
Point.ZERO = I;
const padh = (num, pad) => num.toString(16).padStart(pad, '0');
const bytesToHex = (b) => Array.from(abytes(b))
    .map((e) => padh(e, 2))
    .join('');
const C = { _0: 48, _9: 57, A: 65, F: 70, a: 97, f: 102 };
const _ch = (ch) => {
    if (ch >= C._0 && ch <= C._9)
        return ch - C._0;
    if (ch >= C.A && ch <= C.F)
        return ch - (C.A - 10);
    if (ch >= C.a && ch <= C.f)
        return ch - (C.a - 10);
    return;
};
const hexToBytes = (hex) => {
    const e = 'hex invalid';
    if (!isS(hex))
        return err(e);
    const hl = hex.length, al = hl / 2;
    if (hl % 2)
        return err(e);
    const array = u8n(al);
    for (let ai = 0, hi = 0; ai < al; ai++, hi += 2) {
        const n1 = _ch(hex.charCodeAt(hi));
        const n2 = _ch(hex.charCodeAt(hi + 1));
        if (n1 === undefined || n2 === undefined)
            return err(e);
        array[ai] = n1 * 16 + n2;
    }
    return array;
};
const numToBytes_32LE = (num) => hexToBytes(padh(num, L2)).reverse();
const bytesToNum_LE = (b) => BigInt('0x' + bytesToHex(u8fr(abytes(b)).reverse()));
const concatBytes = (...arrs) => {
    const r = u8n(arrs.reduce((sum, a) => sum + abytes(a).length, 0));
    let pad = 0;
    arrs.forEach((a) => {
        r.set(a, pad);
        pad += a.length;
    });
    return r;
};
const invert = (num, md) => {
    if (num === 0n || md <= 0n)
        err('no inverse n=' + num + ' mod=' + md);
    let a = M(num, md), b = md, x = 0n, y = 1n, u = 1n, v = 0n;
    while (a !== 0n) {
        const q = b / a, r = b % a;
        const m = x - u * q, n = y - v * q;
        (b = a), (a = r), (x = u), (y = v), (u = m), (v = n);
    }
    return b === 1n ? M(x, md) : err('no inverse');
};
const pow2 = (x, power) => {
    let r = x;
    while (power-- > 0n) {
        r *= r;
        r %= P;
    }
    return r;
};
const pow_2_252_3 = (x) => {
    const x2 = (x * x) % P;
    const b2 = (x2 * x) % P;
    const b4 = (pow2(b2, 2n) * b2) % P;
    const b5 = (pow2(b4, 1n) * x) % P;
    const b10 = (pow2(b5, 5n) * b5) % P;
    const b20 = (pow2(b10, 10n) * b10) % P;
    const b40 = (pow2(b20, 20n) * b20) % P;
    const b80 = (pow2(b40, 40n) * b40) % P;
    const b160 = (pow2(b80, 80n) * b80) % P;
    const b240 = (pow2(b160, 80n) * b80) % P;
    const b250 = (pow2(b240, 10n) * b10) % P;
    const pow_p_5_8 = (pow2(b250, 2n) * x) % P;
    return { pow_p_5_8, b2 };
};
const RM1 = 0x2b8324804fc1df0b2b4d00993dfbd7a72f431806ad2fe478c4ee1b274a0ea0b0n;
const uvRatio = (u, v) => {
    const v3 = M(v * v * v);
    const v7 = M(v3 * v3 * v);
    const pow = pow_2_252_3(u * v7).pow_p_5_8;
    let x = M(u * v3 * pow);
    const vx2 = M(v * x * x);
    const root1 = x;
    const root2 = M(x * RM1);
    const useRoot1 = vx2 === u;
    const useRoot2 = vx2 === M(-u);
    const noRoot = vx2 === M(-u * RM1);
    if (useRoot1)
        x = root1;
    if (useRoot2 || noRoot)
        x = root2;
    if ((M(x) & 1n) === 1n)
        x = M(-x);
    return { isValid: useRoot1 || useRoot2, value: x };
};
const modL_LE = (hash) => M(bytesToNum_LE(hash), N);
const sha512a = (...m) => etc.sha512Async(...m);
const sha512s = (...m) => {
    const fn = etc.sha512Sync;
    if (typeof fn !== 'function')
        err('etc.sha512Sync not set');
    return fn(...m);
};
const hash2extK = (hashed) => {
    const head = hashed.slice(0, L);
    head[0] &= 248;
    head[31] &= 127;
    head[31] |= 64;
    const prefix = hashed.slice(L, L2);
    const scalar = modL_LE(head);
    const point = G.mul(scalar);
    const pointBytes = point.toBytes();
    return { head, prefix, scalar, point, pointBytes };
};
const getExtendedPublicKeyAsync = (priv) => sha512a(toU8(priv, L)).then(hash2extK);
const getExtendedPublicKey = (priv) => hash2extK(sha512s(toU8(priv, L)));
const getPublicKeyAsync = (priv) => getExtendedPublicKeyAsync(priv).then((p) => p.pointBytes);
exports.getPublicKeyAsync = getPublicKeyAsync;
const getPublicKey = (priv) => getExtendedPublicKey(priv).pointBytes;
exports.getPublicKey = getPublicKey;
const hashFinishA = (res) => sha512a(res.hashable).then(res.finish);
const hashFinishS = (res) => res.finish(sha512s(res.hashable));
const _sign = (e, rBytes, msg) => {
    const { pointBytes: P, scalar: s } = e;
    const r = modL_LE(rBytes);
    const R = G.mul(r).toBytes();
    const hashable = concatBytes(R, P, msg);
    const finish = (hashed) => {
        const S = M(r + modL_LE(hashed) * s, N);
        return abytes(concatBytes(R, numToBytes_32LE(S)), L2);
    };
    return { hashable, finish };
};
const signAsync = async (msg, privKey) => {
    const m = toU8(msg);
    const e = await getExtendedPublicKeyAsync(privKey);
    const rBytes = await sha512a(e.prefix, m);
    return hashFinishA(_sign(e, rBytes, m));
};
exports.signAsync = signAsync;
const sign = (msg, privKey) => {
    const m = toU8(msg);
    const e = getExtendedPublicKey(privKey);
    const rBytes = sha512s(e.prefix, m);
    return hashFinishS(_sign(e, rBytes, m));
};
exports.sign = sign;
const dvo = { zip215: true };
const _verify = (sig, msg, pub, opts = dvo) => {
    sig = toU8(sig, L2);
    msg = toU8(msg);
    pub = toU8(pub, L);
    const { zip215 } = opts;
    let A, R, s, SB, hashable = Uint8Array.of();
    try {
        A = Point.fromHex(pub, zip215);
        R = Point.fromHex(sig.slice(0, L), zip215);
        s = bytesToNum_LE(sig.slice(L, L2));
        SB = G.mul(s, false);
        hashable = concatBytes(R.toBytes(), A.toBytes(), msg);
    }
    catch (error) { }
    const finish = (hashed) => {
        if (SB == null)
            return false;
        if (!zip215 && A.isSmallOrder())
            return false;
        const k = modL_LE(hashed);
        const RkA = R.add(A.mul(k, false));
        return RkA.add(SB.negate()).clearCofactor().is0();
    };
    return { hashable, finish };
};
const verifyAsync = async (s, m, p, opts = dvo) => hashFinishA(_verify(s, m, p, opts));
exports.verifyAsync = verifyAsync;
const verify = (s, m, p, opts = dvo) => hashFinishS(_verify(s, m, p, opts));
exports.verify = verify;
const cr = () => crypto_1.webcrypto;
const subtle = () => cr()?.subtle ?? err('crypto.subtle must be defined');
const rand = (len = 32) => {
    const c = cr();
    if (!c?.getRandomValues)
        err('crypto.getRandomValues must be defined');
    return c.getRandomValues(u8n(len));
};
const etc = {
    sha512Async: async (...messages) => {
        const s = subtle();
        const m = concatBytes(...messages);
        return new Uint8Array(await s.digest('SHA-512', m));
    },
    sha512Sync: undefined,
    bytesToHex: bytesToHex,
    hexToBytes: hexToBytes,
    concatBytes: concatBytes,
    mod: M,
    invert: invert,
    randomBytes: rand,
};
exports.etc = etc;
const utils = {
    getExtendedPublicKeyAsync: getExtendedPublicKeyAsync,
    getExtendedPublicKey: getExtendedPublicKey,
    randomPrivateKey: () => rand(L),
    precompute: (w = 8, p = G) => {
        p.multiply(3n);
        w;
        return p;
    },
};
exports.utils = utils;
const W = 8;
const scalarBits = 256;
const pwindows = Math.ceil(scalarBits / W) + 1;
const pwindowSize = 2 ** (W - 1);
const precompute = () => {
    const points = [];
    let p = G, b = p;
    for (let w = 0; w < pwindows; w++) {
        b = p;
        points.push(b);
        for (let i = 1; i < pwindowSize; i++) {
            b = b.add(p);
            points.push(b);
        }
        p = b.double();
    }
    return points;
};
let Gpows = undefined;
const wNAF = (n) => {
    const comp = Gpows || (Gpows = precompute());
    const ctneg = (cnd, p) => {
        let n = p.negate();
        return cnd ? n : p;
    };
    let p = I, f = G;
    const pow_2_w = 2 ** W;
    const maxNum = pow_2_w;
    const mask = BigInt(pow_2_w - 1);
    const shiftBy = BigInt(W);
    for (let w = 0; w < pwindows; w++) {
        let wbits = Number(n & mask);
        n >>= shiftBy;
        if (wbits > pwindowSize) {
            wbits -= maxNum;
            n += 1n;
        }
        const off = w * pwindowSize;
        const offF = off, offP = off + Math.abs(wbits) - 1;
        const isEven = w % 2 !== 0, isNeg = wbits < 0;
        if (wbits === 0) {
            f = f.add(ctneg(isEven, comp[offF]));
        }
        else {
            p = p.add(ctneg(isNeg, comp[offP]));
        }
    }
    return { p, f };
};
