// n8n-nodes-1shot/dist/nodes/OneShot/OneShotSynch.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OneShotSynch = void 0;
const constants_1 = require("./types/constants");
const ContractMethods_1 = require("./executions/ContractMethods");
const options_1 = require("./executions/options");
const Wallets_1 = require("./executions/Wallets");
class OneShotSynch {
    constructor() {
        this.description = {
            displayName: '1Shot API Submit & Wait',
            name: 'oneShotSynch',
            icon: { light: 'file:oneshot.svg', dark: 'file:oneshot.svg' },
            group: ['transform'],
            version: 1,
            subtitle: 'Execute and wait for transaction completion',
            description: 'Execute a contract method and wait for the transaction to complete',
            defaults: {
                name: '1Shot API Submit & Wait',
            },
            inputs: ["main"],
            outputs: ["main", "main"],
            outputNames: ['Success', 'Error'],
            usableAsTool: true,
            credentials: [
                {
                    name: 'oneShotOAuth2Api',
                    required: true,
                },
            ],
            requestDefaults: {
                baseURL: constants_1.oneshotApiBaseUrl,
                headers: {
                    Accept: 'application/json',
                    'Content-Type': 'application/json',
                },
            },
            properties: [
                {
                    displayName: 'Operation',
                    name: 'operation',
                    type: 'options',
                    noDataExpression: true,
                    displayOptions: {},
                    options: [
                        {
                            name: 'Execute',
                            value: 'execute',
                            description: 'Execute a contract method on the blockchain',
                            action: 'Execute a contract method',
                        },
                        {
                            name: 'Execute as Delegator',
                            value: 'executeAsDelegator',
                            description: 'Execute a contract method on the blockchain using a stored ERC-7702 delgation',
                            action: 'Execute a contract method as delegator',
                        },
                        {
                            name: 'Execute Batch',
                            value: 'executeBatch',
                            description: 'Execute multiple smart contract methods in a single transaction',
                            action: 'Execute multiple smart contract methods',
                        },
                        {
                            name: 'Execute Batch as Delegator',
                            value: 'executeAsDelegatorBatch',
                            description: 'Execute multiple smart contract methods as a delegator in a single transaction',
                            action: 'Execute multiple smart contract methods as delegator',
                        },
                    ],
                    default: 'execute',
                },
                {
                    displayName: 'Contract Method Name or ID',
                    name: 'contractMethodId',
                    type: 'options',
                    typeOptions: {
                        loadOptionsMethod: 'loadContractMethodExecutionOptions',
                    },
                    required: true,
                    displayOptions: {
                        show: {
                            operation: ['execute', 'executeAsDelegator'],
                        },
                    },
                    default: '',
                    description: 'Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code/expressions/">expression</a>',
                },
                {
                    displayName: 'Parameters',
                    name: 'params',
                    type: 'json',
                    required: true,
                    displayOptions: {
                        show: {
                            operation: ['execute', 'executeAsDelegator'],
                        },
                    },
                    default: '{}',
                    description: 'The parameters to pass to the Contract Method. Enter a JSON object (e.g., {"to": "0x3e6a2f0CBA03d293B54c9fCF354948903007a798", "amount": "10000"}).',
                },
                {
                    displayName: 'Delegator Wallet Address',
                    name: 'delegatorWalletAddress',
                    type: 'string',
                    required: true,
                    displayOptions: {
                        show: {
                            operation: ['executeAsDelegator'],
                        },
                    },
                    default: '',
                    description: 'The address of the delagator wallet to use for the contract method',
                },
                {
                    displayName: 'Contract Methods',
                    name: 'contractMethods',
                    type: 'fixedCollection',
                    typeOptions: {
                        multipleValues: true,
                    },
                    required: true,
                    displayOptions: {
                        show: {
                            operation: ['executeBatch'],
                        },
                    },
                    default: {},
                    options: [
                        {
                            name: 'contractMethod',
                            displayName: 'Contract Method',
                            values: [
                                {
                                    displayName: 'Contract Method Name or ID',
                                    name: 'contractMethodId',
                                    type: 'options',
                                    typeOptions: {
                                        loadOptionsMethod: 'loadContractMethodExecutionOptions',
                                    },
                                    required: true,
                                    default: '',
                                    description: 'Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code/expressions/">expression</a>',
                                },
                                {
                                    displayName: 'Parameters',
                                    name: 'params',
                                    type: 'json',
                                    required: true,
                                    default: '{}',
                                    description: 'The parameters to pass to the Contract Method. Enter a JSON object (e.g., {"to": "0x3e6a2f0CBA03d293B54c9fCF354948903007a798", "amount": "10000"}).',
                                },
                                {
                                    displayName: 'Optional Fields',
                                    name: 'optionalFields',
                                    type: 'collection',
                                    placeholder: 'Add Field',
                                    default: {},
                                    options: [
                                        {
                                            displayName: 'Contract Address Override',
                                            name: 'contractAddress',
                                            type: 'string',
                                            default: '',
                                            description: 'Set this to override the default contract address for this method execution',
                                        },
                                        {
                                            displayName: 'Value',
                                            name: 'value',
                                            type: 'string',
                                            default: '',
                                            description: 'The amount of native token to send along with the Contract Method',
                                        },
                                    ],
                                },
                            ],
                        },
                    ],
                },
                {
                    displayName: 'Contract Methods',
                    name: 'contractMethods',
                    type: 'fixedCollection',
                    typeOptions: {
                        multipleValues: true,
                    },
                    required: true,
                    displayOptions: {
                        show: {
                            operation: ['executeAsDelegatorBatch'],
                        },
                    },
                    default: {},
                    options: [
                        {
                            name: 'contractMethod',
                            displayName: 'Contract Method',
                            values: [
                                {
                                    displayName: 'Contract Method Name or ID',
                                    name: 'contractMethodId',
                                    type: 'options',
                                    typeOptions: {
                                        loadOptionsMethod: 'loadContractMethodExecutionOptions',
                                    },
                                    required: true,
                                    default: '',
                                    description: 'Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code/expressions/">expression</a>',
                                },
                                {
                                    displayName: 'Parameters',
                                    name: 'params',
                                    type: 'json',
                                    required: true,
                                    default: '{}',
                                    description: 'The parameters to pass to the Contract Method. Enter a JSON object (e.g., {"to": "0x3e6a2f0CBA03d293B54c9fCF354948903007a798", "amount": "10000"}).',
                                },
                                {
                                    displayName: 'Delegator Address',
                                    name: 'delegatorAddress',
                                    type: 'string',
                                    required: true,
                                    default: '',
                                    description: 'The address of the delegator on whose behalf the transaction will be executed',
                                },
                                {
                                    displayName: 'Optional Fields',
                                    name: 'optionalFields',
                                    type: 'collection',
                                    placeholder: 'Add Field',
                                    default: {},
                                    options: [
                                        {
                                            displayName: 'Contract Address Override',
                                            name: 'contractAddress',
                                            type: 'string',
                                            default: '',
                                            description: 'Set this to override the default contract address for this method execution',
                                        },
                                        {
                                            displayName: 'Value',
                                            name: 'value',
                                            type: 'string',
                                            default: '',
                                            description: 'The amount of native token to send along with the Contract Method',
                                        },
                                    ],
                                },
                            ],
                        },
                    ],
                },
                {
                    displayName: 'Wallet Name or ID',
                    name: 'walletId',
                    type: 'options',
                    typeOptions: {
                        loadOptionsMethod: 'loadWalletOptions',
                    },
                    required: true,
                    displayOptions: {
                        show: {
                            operation: ['executeBatch', 'executeAsDelegatorBatch'],
                        },
                    },
                    default: '',
                    description: 'Select the Wallet to use for the contract methods. Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code/expressions/">expression</a>.',
                },
                {
                    displayName: 'Additional Fields',
                    name: 'additionalFields',
                    type: 'collection',
                    placeholder: 'Add Field',
                    default: {},
                    displayOptions: {
                        show: {
                            operation: ['execute', 'executeAsDelegator'],
                        },
                    },
                    options: [
                        {
                            displayName: 'Wallet ID',
                            name: 'walletId',
                            type: 'string',
                            default: '',
                            description: 'The ID of the Wallet to use for this Contract Method',
                        },
                        {
                            displayName: 'Memo',
                            name: 'memo',
                            type: 'string',
                            default: '',
                            description: 'Optional text to include with the Transaction after the Contract Method is executed',
                        },
                        {
                            displayName: 'Contract Address Override',
                            name: 'contractAddress',
                            type: 'string',
                            default: '',
                            description: 'Set this to override the default contract address for this method execution',
                        },
                        {
                            displayName: 'Value',
                            name: 'value',
                            type: 'string',
                            default: '',
                            description: 'The amount of native token to send along with the Contract Method',
                        },
                        {
                            displayName: 'Authorization List',
                            name: 'authorizationList',
                            type: 'json',
                            default: '[]',
                            description: 'List of ERC-7702 authorizations for the Contract Method',
                        },
                        {
                            displayName: 'Gas Limit',
                            name: 'gasLimit',
                            type: 'string',
                            default: '',
                            description: 'The gas limit to use for the Contract Method. If not provided, the default gas limit will be used.',
                        },
                    ],
                },
                {
                    displayName: 'Additional Fields',
                    name: 'additionalFields',
                    type: 'collection',
                    placeholder: 'Add Field',
                    default: {},
                    displayOptions: {
                        show: {
                            operation: ['executeBatch', 'executeAsDelegatorBatch'],
                        },
                    },
                    options: [
                        {
                            displayName: 'Atomic',
                            name: 'atomic',
                            type: 'boolean',
                            default: true,
                            description: 'Whether or not all the transactions in the batch must succeed, or the entire batch is rolled back',
                        },
                        {
                            displayName: 'Memo',
                            name: 'memo',
                            type: 'string',
                            default: '',
                            description: 'Optional text to include with the Transaction after the batch is executed',
                        },
                        {
                            displayName: 'Authorization List',
                            name: 'authorizationList',
                            type: 'json',
                            default: '[]',
                            description: 'List of ERC-7702 authorizations for the batch execution',
                        },
                        {
                            displayName: 'Gas Limit',
                            name: 'gasLimit',
                            type: 'string',
                            default: '',
                            description: 'The gas limit to use for the batch execution. The higher of either this value or the estimated gas will be used.',
                        },
                    ],
                },
            ],
        };
        this.methods = {
            loadOptions: {
                loadContractMethodExecutionOptions: options_1.loadContractMethodExecutionOptions,
                loadWalletOptions: Wallets_1.loadWalletOptions,
            },
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        const errorData = [];
        try {
            for (let i = 0; i < items.length; i++) {
                const operation = this.getNodeParameter('operation', i);
                if (operation === 'execute') {
                    const response = await (0, ContractMethods_1.executeAndWaitContractMethodOperation)(this, i);
                    if (response.success) {
                        returnData.push({
                            json: { ...response.result },
                            pairedItem: { item: i },
                        });
                    }
                    else {
                        errorData.push({
                            json: { ...response.result },
                            pairedItem: { item: i },
                        });
                    }
                }
                else if (operation === 'executeAsDelegator') {
                    const response = await (0, ContractMethods_1.executeAsDelegatorAndWaitContractMethodOperation)(this, i);
                    if (response.success) {
                        returnData.push({
                            json: { ...response.result },
                            pairedItem: { item: i },
                        });
                    }
                    else {
                        errorData.push({
                            json: { ...response.result },
                            pairedItem: { item: i },
                        });
                    }
                }
                else if (operation === 'executeBatch') {
                    const response = await (0, ContractMethods_1.executeBatchAndWaitOperation)(this, i);
                    if (response.success) {
                        returnData.push({
                            json: { ...response.result },
                            pairedItem: { item: i },
                        });
                    }
                    else {
                        errorData.push({
                            json: { ...response.result },
                            pairedItem: { item: i },
                        });
                    }
                }
                else if (operation === 'executeAsDelegatorBatch') {
                    const response = await (0, ContractMethods_1.executeAsDelegatorBatchAndWaitOperation)(this, i);
                    if (response.success) {
                        returnData.push({
                            json: { ...response.result },
                            pairedItem: { item: i },
                        });
                    }
                    else {
                        errorData.push({
                            json: { ...response.result },
                            pairedItem: { item: i },
                        });
                    }
                }
            }
        }
        catch (error) {
            errorData.push({ json: { ...error }, pairedItem: { item: 0 } });
        }
        return [returnData, errorData];
    }
}
exports.OneShotSynch = OneShotSynch;
