// n8n-nodes-1shot/dist/nodes/OneShot/OneShot.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OneShot = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const ContractMethodDescription_1 = require("./descriptions/ContractMethodDescription");
const ContractEventDescription_1 = require("./descriptions/ContractEventDescription");
const WalletDescription_1 = require("./descriptions/WalletDescription");
const PromptDescription_1 = require("./descriptions/PromptDescription");
const TransactionDescription_1 = require("./descriptions/TransactionDescription");
const options_1 = require("./executions/options");
const Wallets_1 = require("./executions/Wallets");
const ContractMethods_1 = require("./executions/ContractMethods");
const ContractEvents_1 = require("./executions/ContractEvents");
const constants_1 = require("./types/constants");
const Transactions_1 = require("./executions/Transactions");
const Prompts_1 = require("./executions/Prompts");
const ChainDescription_1 = require("./descriptions/ChainDescription");
const Chains_1 = require("./executions/Chains");
const X402RequestDescription_1 = require("./descriptions/X402RequestDescription");
const genericRequestFunctions_1 = require("./utils/genericRequestFunctions");
const n8nUtils_1 = require("./utils/n8nUtils");
const lodashFunctions_1 = require("./utils/lodashFunctions");
const binaryData_1 = require("./utils/binaryData");
const parse_1 = require("./utils/parse");
class OneShot {
    constructor() {
        this.description = {
            displayName: '1Shot API',
            name: 'oneShot',
            icon: { light: 'file:oneshot.svg', dark: 'file:oneshot.svg' },
            group: ['transform'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Interact with EVM blockchains via 1Shot API',
            defaults: {
                name: '1Shot API',
            },
            inputs: ["main"],
            outputs: ["main"],
            usableAsTool: true,
            credentials: [
                {
                    name: 'oneShotOAuth2Api',
                    required: true,
                },
            ],
            requestDefaults: {
                baseURL: constants_1.oneshotApiBaseUrl,
                url: '',
                headers: {
                    Accept: 'application/json',
                    'Content-Type': 'application/json',
                },
            },
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'X402 Request',
                            value: 'x402Request',
                        },
                        {
                            name: 'Chain',
                            value: 'chains',
                        },
                        {
                            name: 'Contract Event',
                            value: 'contractEvents',
                        },
                        {
                            name: 'Contract Method',
                            value: 'contractMethods',
                        },
                        {
                            name: 'Prompt',
                            value: 'prompts',
                        },
                        {
                            name: 'Transaction',
                            value: 'transactions',
                        },
                        {
                            name: 'Wallet',
                            value: 'wallets',
                        },
                    ],
                    default: 'contractMethods',
                },
                ...ChainDescription_1.chainOperationsFields,
                ...ContractEventDescription_1.contractEventOperationsFields,
                ...ContractMethodDescription_1.contractMethodOperationsFields,
                ...WalletDescription_1.walletOperationsFields,
                ...PromptDescription_1.promptOperationsFields,
                ...TransactionDescription_1.transactionOperationsFields,
                ...X402RequestDescription_1.x402RequestProperties,
            ],
        };
        this.methods = {
            loadOptions: {
                loadChainOptions: options_1.loadChainOptions,
                loadChainOptionsWithAll: options_1.loadChainOptionsWithAll,
                loadContractMethodExecutionOptions: options_1.loadContractMethodExecutionOptions,
                loadContractMethodReadOptions: options_1.loadContractMethodReadOptions,
                loadContractMethodAllOptions: options_1.loadContractMethodAllOptions,
                loadContractEventOptions: options_1.loadContractEventOptions,
                loadWalletOptions: Wallets_1.loadWalletOptions,
            },
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        if (this.getNodeParameter('resource') === 'x402Request') {
            return await executeX402RequestOperation.call(this);
        }
        for (let i = 0; i < items.length; i++) {
            const resource = this.getNodeParameter('resource', i);
            const operation = this.getNodeParameter('operation', i);
            if (resource === 'contractMethods') {
                if (operation === 'execute') {
                    const response = await (0, ContractMethods_1.executeContractMethodOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'executeAsDelegator') {
                    const response = await (0, ContractMethods_1.executeAsDelegatorContractMethodOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'encode') {
                    const response = await (0, ContractMethods_1.encodeContractMethodOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'estimate') {
                    const response = await (0, ContractMethods_1.estimateContractMethodOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'simulate') {
                    const response = await (0, ContractMethods_1.simulateContractMethodOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'read') {
                    const response = await (0, ContractMethods_1.readContractMethodOperation)(this, i);
                    returnData.push({ json: { response: response }, pairedItem: { item: i } });
                }
                else if (operation === 'list') {
                    const response = await (0, ContractMethods_1.listContractMethodsOperation)(this, i);
                    returnData.push(...response.response.map((item) => ({
                        json: item,
                        pairedItem: { item: i },
                    })));
                }
                else if (operation === 'get') {
                    const response = await (0, ContractMethods_1.getContractMethodOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'assureContractMethodsFromPrompt') {
                    const response = await (0, ContractMethods_1.assureContractMethodsFromPromptOperation)(this, i);
                    returnData.push(...response.map((item) => ({
                        json: item,
                        pairedItem: { item: i },
                    })));
                }
                else if (operation === 'executeBatch') {
                    const response = await (0, ContractMethods_1.executeBatchOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'executeAsDelegatorBatch') {
                    const response = await (0, ContractMethods_1.executeAsDelegatorBatchOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Unsupported operation for resource contractMethods: ${operation}`);
                }
            }
            else if (resource === 'contractEvents') {
                if (operation === 'list') {
                    const response = await (0, ContractEvents_1.listContractEventsOperation)(this, i);
                    returnData.push(...response.response.map((item) => ({
                        json: item,
                        pairedItem: { item: i },
                    })));
                }
                else if (operation === 'create') {
                    const response = await (0, ContractEvents_1.createContractEventOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'get') {
                    const response = await (0, ContractEvents_1.getContractEventOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'update') {
                    const response = await (0, ContractEvents_1.updateContractEventOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'delete') {
                    const response = await (0, ContractEvents_1.deleteContractEventOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'search') {
                    const response = await (0, ContractEvents_1.searchContractEventOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Unsupported operation for resource contractEvents: ${operation}`);
                }
            }
            else if (resource === 'wallets') {
                if (operation === 'list') {
                    const response = await (0, Wallets_1.listWalletsOperation)(this, i);
                    returnData.push(...response.response.map((item) => ({
                        json: item,
                        pairedItem: { item: i },
                    })));
                }
                else if (operation === 'create') {
                    const response = await (0, Wallets_1.createWalletOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'get') {
                    const response = await (0, Wallets_1.getWalletOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'update') {
                    const response = await (0, Wallets_1.updateWalletOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else if (operation === 'delete') {
                    const response = await (0, Wallets_1.deleteWalletOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Unsupported operation for resource wallets: ${operation}`);
                }
            }
            else if (resource === 'prompts') {
                if (operation === 'search') {
                    const response = await (0, Prompts_1.searchPromptsOperation)(this, i);
                    returnData.push(...response.map((item) => ({
                        json: item,
                        pairedItem: { item: i },
                    })));
                }
                else {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Unsupported operation for resource prompts: ${operation}`);
                }
            }
            else if (resource === 'transactions') {
                if (operation === 'list') {
                    const response = await (0, Transactions_1.listTransactionsOperation)(this, i);
                    returnData.push(...response.response.map((item) => ({
                        json: item,
                        pairedItem: { item: i },
                    })));
                }
                else if (operation === 'get') {
                    const response = await (0, Transactions_1.getTransactionOperation)(this, i);
                    returnData.push({ json: response, pairedItem: { item: i } });
                }
                else {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Unsupported operation for resource transactions: ${operation}`);
                }
            }
            else if (resource === 'chains') {
                if (operation === 'list') {
                    const response = await (0, Chains_1.listChainsOperation)(this, i);
                    returnData.push(...response.response.map((item) => ({
                        json: item,
                        pairedItem: { item: i },
                    })));
                }
                else {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Unsupported operation for resource chains: ${operation}`);
                }
            }
            else {
                throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Unsupported resource: ${resource}`);
            }
        }
        return [returnData];
    }
}
exports.OneShot = OneShot;
async function executeX402RequestOperation() {
    const items = this.getInputData();
    const nodeVersion = this.getNode().typeVersion;
    const fullResponseProperties = ['body', 'headers', 'statusCode', 'statusMessage'];
    let authentication;
    try {
        authentication = this.getNodeParameter('authentication', 0);
    }
    catch { }
    let httpBasicAuth;
    let httpBearerAuth;
    let httpDigestAuth;
    let httpHeaderAuth;
    let httpQueryAuth;
    let httpCustomAuth;
    let oAuth1Api;
    let oAuth2Api;
    let sslCertificates;
    let nodeCredentialType;
    let genericCredentialType;
    let requestOptions = {
        url: '',
    };
    let returnItems = [];
    const errorItems = {};
    const requestPromises = [];
    let fullResponse = false;
    let autoDetectResponseFormat = false;
    let responseFileName;
    const requests = [];
    const updadeQueryParameter = (0, genericRequestFunctions_1.updadeQueryParameterConfig)(nodeVersion);
    for (let itemIndex = 0; itemIndex < items.length; itemIndex++) {
        try {
            if (authentication === 'genericCredentialType') {
                genericCredentialType = this.getNodeParameter('genericAuthType', 0);
                if (genericCredentialType === 'httpBasicAuth') {
                    httpBasicAuth = await this.getCredentials('httpBasicAuth', itemIndex);
                }
                else if (genericCredentialType === 'httpBearerAuth') {
                    httpBearerAuth = await this.getCredentials('httpBearerAuth', itemIndex);
                }
                else if (genericCredentialType === 'httpDigestAuth') {
                    httpDigestAuth = await this.getCredentials('httpDigestAuth', itemIndex);
                }
                else if (genericCredentialType === 'httpHeaderAuth') {
                    httpHeaderAuth = await this.getCredentials('httpHeaderAuth', itemIndex);
                }
                else if (genericCredentialType === 'httpQueryAuth') {
                    httpQueryAuth = await this.getCredentials('httpQueryAuth', itemIndex);
                }
                else if (genericCredentialType === 'httpCustomAuth') {
                    httpCustomAuth = await this.getCredentials('httpCustomAuth', itemIndex);
                }
                else if (genericCredentialType === 'oAuth1Api') {
                    oAuth1Api = await this.getCredentials('oAuth1Api', itemIndex);
                }
                else if (genericCredentialType === 'oAuth2Api') {
                    oAuth2Api = await this.getCredentials('oAuth2Api', itemIndex);
                }
            }
            else if (authentication === 'predefinedCredentialType') {
                nodeCredentialType = this.getNodeParameter('nodeCredentialType', itemIndex);
            }
            const url = this.getNodeParameter('url', itemIndex);
            if (typeof url !== 'string') {
                const actualType = url === null ? 'null' : typeof url;
                throw new n8n_workflow_1.NodeOperationError(this.getNode(), `URL parameter must be a string, got ${actualType}`);
            }
            if (!url.startsWith('http://') && !url.startsWith('https://')) {
                throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Invalid URL: ${url}. URL must start with "http" or "https".`);
            }
            const checkDomainRestrictions = async (credentialData, url, credentialType) => {
                if (credentialData.allowedHttpRequestDomains === 'domains') {
                    const allowedDomains = credentialData.allowedDomains;
                    if (!allowedDomains || allowedDomains.trim() === '') {
                        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'No allowed domains specified. Configure allowed domains or change restriction setting.');
                    }
                    if (!(0, n8nUtils_1.isDomainAllowed)(url, { allowedDomains })) {
                        const credentialInfo = credentialType ? ` (${credentialType})` : '';
                        throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Domain not allowed: This credential${credentialInfo} is restricted from accessing ${url}. ` +
                            `Only the following domains are allowed: ${allowedDomains}`);
                    }
                }
                else if (credentialData.allowedHttpRequestDomains === 'none') {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'This credential is configured to prevent use within an HTTP Request node');
                }
            };
            if (httpBasicAuth)
                await checkDomainRestrictions(httpBasicAuth, url);
            if (httpBearerAuth)
                await checkDomainRestrictions(httpBearerAuth, url);
            if (httpDigestAuth)
                await checkDomainRestrictions(httpDigestAuth, url);
            if (httpHeaderAuth)
                await checkDomainRestrictions(httpHeaderAuth, url);
            if (httpQueryAuth)
                await checkDomainRestrictions(httpQueryAuth, url);
            if (httpCustomAuth)
                await checkDomainRestrictions(httpCustomAuth, url);
            if (oAuth1Api)
                await checkDomainRestrictions(oAuth1Api, url);
            if (oAuth2Api)
                await checkDomainRestrictions(oAuth2Api, url);
            if (nodeCredentialType) {
                try {
                    const credentialData = await this.getCredentials(nodeCredentialType, itemIndex);
                    await checkDomainRestrictions(credentialData, url, nodeCredentialType);
                }
                catch (error) {
                    if (error.message?.includes('Domain not allowed') ||
                        error.message?.includes('configured to prevent') ||
                        error.message?.includes('No allowed domains specified')) {
                        throw error;
                    }
                }
            }
            const provideSslCertificates = this.getNodeParameter('provideSslCertificates', itemIndex, false);
            if (provideSslCertificates) {
                sslCertificates = await this.getCredentials('httpSslAuth', itemIndex);
            }
            const requestMethod = this.getNodeParameter('method', itemIndex);
            const sendQuery = this.getNodeParameter('sendQuery', itemIndex, false);
            const queryParameters = this.getNodeParameter('queryParameters.parameters', itemIndex, []);
            const specifyQuery = this.getNodeParameter('specifyQuery', itemIndex, 'keypair');
            const jsonQueryParameter = this.getNodeParameter('jsonQuery', itemIndex, '');
            const sendBody = this.getNodeParameter('sendBody', itemIndex, false);
            const bodyContentType = this.getNodeParameter('contentType', itemIndex, '');
            const specifyBody = this.getNodeParameter('specifyBody', itemIndex, '');
            const bodyParameters = this.getNodeParameter('bodyParameters.parameters', itemIndex, []);
            const jsonBodyParameter = this.getNodeParameter('jsonBody', itemIndex, '');
            const body = this.getNodeParameter('body', itemIndex, '');
            const sendHeaders = this.getNodeParameter('sendHeaders', itemIndex, false);
            const headerParameters = this.getNodeParameter('headerParameters.parameters', itemIndex, []);
            const specifyHeaders = this.getNodeParameter('specifyHeaders', itemIndex, 'keypair');
            const jsonHeadersParameter = this.getNodeParameter('jsonHeaders', itemIndex, '');
            const { redirect, batching, timeout, allowUnauthorizedCerts, queryParameterArrays, response, lowercaseHeaders, } = this.getNodeParameter('options', itemIndex, {});
            responseFileName = response?.response?.outputPropertyName;
            const responseFormat = response?.response?.responseFormat || 'autodetect';
            fullResponse = response?.response?.fullResponse || false;
            autoDetectResponseFormat = responseFormat === 'autodetect';
            const batchSize = batching?.batch?.batchSize > 0 ? batching?.batch?.batchSize : 1;
            const batchInterval = batching?.batch.batchInterval;
            if (itemIndex > 0 && batchSize >= 0 && batchInterval > 0) {
                if (itemIndex % batchSize === 0) {
                    await (0, n8n_workflow_1.sleep)(batchInterval);
                }
            }
            requestOptions = {
                headers: {},
                method: requestMethod,
                url: url,
                skipSslCertificateValidation: allowUnauthorizedCerts || false,
                disableFollowRedirect: true,
                returnFullResponse: true,
                ignoreHttpStatusErrors: false,
            };
            if (requestOptions.method !== 'GET' && nodeVersion >= 4.1) {
                requestOptions = { ...requestOptions, disableFollowRedirect: true };
            }
            const defaultRedirect = nodeVersion >= 4 && redirect === undefined;
            if (redirect?.redirect?.followRedirects || defaultRedirect) {
                requestOptions.disableFollowRedirect = false;
            }
            if (response?.response?.neverError) {
                requestOptions.ignoreHttpStatusErrors = true;
            }
            if (timeout) {
                requestOptions.timeout = timeout;
            }
            else {
                requestOptions.timeout = 300000;
            }
            if (sendQuery && queryParameterArrays) {
                Object.assign(requestOptions, {
                    qsStringifyOptions: { arrayFormat: queryParameterArrays },
                });
            }
            const parametersToKeyValue = async (accumulator, cur) => {
                if (cur.parameterType === 'formBinaryData') {
                    if (!cur.inputDataFieldName)
                        return accumulator;
                    const binaryData = this.helpers.assertBinaryData(itemIndex, cur.inputDataFieldName);
                    let uploadData;
                    const itemBinaryData = items[itemIndex].binary[cur.inputDataFieldName];
                    if (itemBinaryData.id) {
                        uploadData = await this.helpers.getBinaryStream(itemBinaryData.id);
                    }
                    else {
                        uploadData = Buffer.from(itemBinaryData.data, n8n_workflow_1.BINARY_ENCODING);
                    }
                    accumulator[cur.name] = {
                        value: uploadData,
                        options: {
                            filename: binaryData.fileName,
                            contentType: binaryData.mimeType,
                        },
                    };
                    return accumulator;
                }
                updadeQueryParameter(accumulator, cur.name, cur.value);
                return accumulator;
            };
            if (sendBody && bodyParameters) {
                if (specifyBody === 'keypair' || bodyContentType === 'multipart-form-data') {
                    requestOptions.body = await (0, genericRequestFunctions_1.prepareRequestBody)(bodyParameters, bodyContentType, nodeVersion, parametersToKeyValue);
                }
                else if (specifyBody === 'json') {
                    if (typeof jsonBodyParameter !== 'object' && jsonBodyParameter !== null) {
                        try {
                            JSON.parse(jsonBodyParameter);
                        }
                        catch {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'JSON parameter needs to be valid JSON', {
                                itemIndex,
                            });
                        }
                        requestOptions.body = (0, n8n_workflow_1.jsonParse)(jsonBodyParameter);
                    }
                    else {
                        requestOptions.body = jsonBodyParameter;
                    }
                }
                else if (specifyBody === 'string') {
                    requestOptions.body = Object.fromEntries(new URLSearchParams(body));
                }
            }
            if (sendBody && ['PATCH', 'POST', 'PUT', 'GET'].includes(requestMethod)) {
                if (bodyContentType === 'binaryData') {
                    const inputDataFieldName = this.getNodeParameter('inputDataFieldName', itemIndex);
                    let uploadData;
                    let contentLength;
                    const itemBinaryData = this.helpers.assertBinaryData(itemIndex, inputDataFieldName);
                    if (itemBinaryData.id) {
                        uploadData = await this.helpers.getBinaryStream(itemBinaryData.id);
                        const metadata = await this.helpers.getBinaryMetadata(itemBinaryData.id);
                        contentLength = metadata.fileSize;
                    }
                    else {
                        uploadData = Buffer.from(itemBinaryData.data, n8n_workflow_1.BINARY_ENCODING);
                        contentLength = uploadData.length;
                    }
                    requestOptions.body = uploadData;
                    requestOptions.headers = {
                        ...requestOptions.headers,
                        'content-length': contentLength,
                        'content-type': itemBinaryData.mimeType ?? 'application/octet-stream',
                    };
                }
                else if (bodyContentType === 'raw') {
                    requestOptions.body = body;
                }
            }
            if (sendQuery && queryParameters) {
                if (specifyQuery === 'keypair') {
                    requestOptions.qs = await (0, genericRequestFunctions_1.reduceAsync)(queryParameters, parametersToKeyValue);
                }
                else if (specifyQuery === 'json') {
                    try {
                        JSON.parse(jsonQueryParameter);
                    }
                    catch {
                        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'JSON parameter needs to be valid JSON', {
                            itemIndex,
                        });
                    }
                    requestOptions.qs = (0, n8n_workflow_1.jsonParse)(jsonQueryParameter);
                }
            }
            if (sendHeaders && headerParameters) {
                let additionalHeaders = {};
                if (specifyHeaders === 'keypair') {
                    additionalHeaders = await (0, genericRequestFunctions_1.reduceAsync)(headerParameters.filter((header) => header.name), parametersToKeyValue);
                }
                else if (specifyHeaders === 'json') {
                    try {
                        JSON.parse(jsonHeadersParameter);
                    }
                    catch {
                        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'JSON parameter needs to be valid JSON', {
                            itemIndex,
                        });
                    }
                    additionalHeaders = (0, n8n_workflow_1.jsonParse)(jsonHeadersParameter);
                }
                requestOptions.headers = {
                    ...requestOptions.headers,
                    ...(lowercaseHeaders === undefined || lowercaseHeaders
                        ? (0, n8nUtils_1.keysToLowercase)(additionalHeaders)
                        : additionalHeaders),
                };
            }
            if (autoDetectResponseFormat || responseFormat === 'file') {
                requestOptions.encoding = undefined;
                requestOptions.json = false;
            }
            else if (bodyContentType === 'raw') {
                requestOptions.json = false;
            }
            else {
                requestOptions.json = true;
            }
            if (bodyContentType === 'raw') {
                if (requestOptions.headers === undefined) {
                    requestOptions.headers = {};
                }
                const rawContentType = this.getNodeParameter('rawContentType', itemIndex);
                requestOptions.headers['content-type'] = rawContentType;
            }
            const authDataKeys = {};
            (0, genericRequestFunctions_1.setAgentOptions)(requestOptions, sslCertificates);
            if (httpBasicAuth !== undefined) {
                requestOptions.auth = {
                    username: httpBasicAuth.user,
                    password: httpBasicAuth.password,
                };
                authDataKeys.auth = ['pass'];
            }
            if (httpBearerAuth !== undefined) {
                requestOptions.headers = requestOptions.headers ?? {};
                requestOptions.headers.Authorization = `Bearer ${String(httpBearerAuth.token)}`;
                authDataKeys.headers = ['Authorization'];
            }
            if (httpHeaderAuth !== undefined) {
                requestOptions.headers[httpHeaderAuth.name] = httpHeaderAuth.value;
                authDataKeys.headers = [httpHeaderAuth.name];
            }
            if (httpQueryAuth !== undefined) {
                if (!requestOptions.qs) {
                    requestOptions.qs = {};
                }
                requestOptions.qs[httpQueryAuth.name] = httpQueryAuth.value;
                authDataKeys.qs = [httpQueryAuth.name];
            }
            if (httpDigestAuth !== undefined) {
                requestOptions.auth = {
                    username: httpDigestAuth.user,
                    password: httpDigestAuth.password,
                    sendImmediately: false,
                };
                authDataKeys.auth = ['pass'];
            }
            if (httpCustomAuth !== undefined) {
                const customAuth = (0, n8n_workflow_1.jsonParse)(httpCustomAuth.json || '{}', { errorMessage: 'Invalid Custom Auth JSON' });
                if (customAuth.headers) {
                    requestOptions.headers = { ...requestOptions.headers, ...customAuth.headers };
                    authDataKeys.headers = Object.keys(customAuth.headers);
                }
                if (customAuth.body) {
                    requestOptions.body = { ...requestOptions.body, ...customAuth.body };
                    authDataKeys.body = Object.keys(customAuth.body);
                }
                if (customAuth.qs) {
                    requestOptions.qs = { ...requestOptions.qs, ...customAuth.qs };
                    authDataKeys.qs = Object.keys(customAuth.qs);
                }
            }
            if (requestOptions.headers.accept === undefined) {
                if (responseFormat === 'json') {
                    requestOptions.headers.accept = 'application/json,text/*;q=0.99';
                }
                else if (responseFormat === 'text') {
                    requestOptions.headers.accept =
                        'application/json,text/html,application/xhtml+xml,application/xml,text/*;q=0.9, */*;q=0.1';
                }
                else {
                    requestOptions.headers.accept =
                        'application/json,text/html,application/xhtml+xml,application/xml,text/*;q=0.9, image/*;q=0.8, */*;q=0.7';
                }
            }
            requests.push({
                options: requestOptions,
                authKeys: authDataKeys,
                credentialType: nodeCredentialType,
            });
            const ignoreHttpStatusErrors = requestOptions.ignoreHttpStatusErrors;
            requestOptions.ignoreHttpStatusErrors = true;
            if (authentication === 'genericCredentialType' || authentication === 'none') {
                if (oAuth1Api) {
                    const requestOAuth1 = this.helpers.httpRequestWithAuthentication
                        .call(this, 'oAuth1Api', requestOptions)
                        .then(async (response) => {
                        if (response.statusCode === 402) {
                            const { paymentHeader, headerKey } = await generateX402PaymentHeader.call(this, response.body, response.headers);
                            requestOptions.headers[headerKey] = paymentHeader;
                            requestOptions.ignoreHttpStatusErrors = ignoreHttpStatusErrors;
                            return this.helpers.httpRequestWithAuthentication.call(this, 'oAuth1Api', requestOptions);
                        }
                        return response;
                    });
                    requestPromises.push(requestOAuth1);
                }
                else if (oAuth2Api) {
                    const requestOAuth2 = this.helpers.httpRequestWithAuthentication
                        .call(this, 'oAuth2Api', requestOptions, {
                        oauth2: { tokenType: 'Bearer' },
                    })
                        .then(async (response) => {
                        if (response.statusCode === 402) {
                            const { paymentHeader, headerKey } = await generateX402PaymentHeader.call(this, response.body, response.headers);
                            requestOptions.headers[headerKey] = paymentHeader;
                            requestOptions.ignoreHttpStatusErrors = ignoreHttpStatusErrors;
                            return this.helpers.httpRequestWithAuthentication.call(this, 'oAuth2Api', requestOptions, {
                                oauth2: { tokenType: 'Bearer' },
                            });
                        }
                        return response;
                    });
                    requestPromises.push(requestOAuth2);
                }
                else {
                    const request = this.helpers.httpRequest(requestOptions).then(async (response) => {
                        if (response.statusCode === 402) {
                            const { paymentHeader, headerKey } = await generateX402PaymentHeader.call(this, response.body, response.headers);
                            requestOptions.headers[headerKey] = paymentHeader;
                            requestOptions.ignoreHttpStatusErrors = ignoreHttpStatusErrors;
                            return this.helpers.httpRequest(requestOptions);
                        }
                        return response;
                    });
                    requestPromises.push(request);
                }
            }
            else if (authentication === 'predefinedCredentialType' && nodeCredentialType) {
                const additionalOAuth2Options = (0, genericRequestFunctions_1.getOAuth2AdditionalParameters)(nodeCredentialType);
                const requestWithAuthentication = this.helpers.httpRequestWithAuthentication
                    .call(this, nodeCredentialType, requestOptions, additionalOAuth2Options && { oauth2: additionalOAuth2Options })
                    .then(async (response) => {
                    if (response.statusCode === 402) {
                        const { paymentHeader, headerKey } = await generateX402PaymentHeader.call(this, response.body, response.headers);
                        requestOptions.headers[headerKey] = paymentHeader;
                        requestOptions.ignoreHttpStatusErrors = ignoreHttpStatusErrors;
                        return this.helpers.httpRequestWithAuthentication.call(this, nodeCredentialType, requestOptions, additionalOAuth2Options && { oauth2: additionalOAuth2Options });
                    }
                    return response;
                });
                requestPromises.push(requestWithAuthentication);
            }
        }
        catch (error) {
            if (!this.continueOnFail())
                throw error;
            requestPromises.push(Promise.reject(error).catch(() => { }));
            errorItems[itemIndex] = error.message;
            continue;
        }
    }
    const sanitizedRequests = [];
    const promisesResponses = await Promise.allSettled(requestPromises.map(async (requestPromise, itemIndex) => await requestPromise
        .then((response) => response)
        .finally(async () => {
        if (errorItems[itemIndex])
            return;
        try {
            const { options, authKeys, credentialType } = requests[itemIndex];
            let secrets = [];
            if (credentialType) {
                const properties = this.getCredentialsProperties(credentialType);
                const credentials = await this.getCredentials(credentialType, itemIndex);
                secrets = (0, genericRequestFunctions_1.getSecrets)(properties, credentials);
            }
            const sanitizedRequestOptions = (0, genericRequestFunctions_1.sanitizeUiMessage)(options, authKeys, secrets);
            sanitizedRequests.push(sanitizedRequestOptions);
            this.sendMessageToUI(sanitizedRequestOptions);
        }
        catch (e) { }
    })));
    let responseData;
    for (let itemIndex = 0; itemIndex < items.length; itemIndex++) {
        try {
            responseData = promisesResponses.shift();
            if (errorItems[itemIndex]) {
                returnItems.push({
                    json: { error: errorItems[itemIndex] },
                    pairedItem: { item: itemIndex },
                });
                continue;
            }
            if (responseData.status !== 'fulfilled') {
                if (responseData.reason.statusCode === 429) {
                    responseData.reason.message =
                        "Try spacing your requests out using the batching settings under 'Options'";
                }
                if (!this.continueOnFail()) {
                    if (autoDetectResponseFormat && responseData.reason.error instanceof Buffer) {
                        responseData.reason.error = Buffer.from(responseData.reason.error).toString();
                    }
                    let error;
                    if (responseData?.reason instanceof n8n_workflow_1.NodeApiError) {
                        error = responseData.reason;
                        (0, lodashFunctions_1.setNestedProperty)(error, 'context.itemIndex', itemIndex);
                    }
                    else {
                        const errorData = (responseData.reason ? responseData.reason : responseData);
                        error = new n8n_workflow_1.NodeApiError(this.getNode(), errorData, { itemIndex });
                    }
                    (0, lodashFunctions_1.setNestedProperty)(error, 'context.request', sanitizedRequests[itemIndex]);
                    throw error;
                }
                else {
                    (0, n8n_workflow_1.removeCircularRefs)(responseData.reason);
                    returnItems.push({
                        json: {
                            error: responseData.reason,
                        },
                        pairedItem: {
                            item: itemIndex,
                        },
                    });
                    continue;
                }
            }
            let responses;
            if (Array.isArray(responseData.value)) {
                responses = responseData.value;
            }
            else {
                responses = [responseData.value];
            }
            let responseFormat = this.getNodeParameter('options.response.response.responseFormat', 0, 'autodetect');
            fullResponse = this.getNodeParameter('options.response.response.fullResponse', 0, false);
            for (let [index, response] of Object.entries(responses)) {
                if (this.getMode() === 'manual' && index === '0') {
                    const nodeContext = this.getContext('node');
                    nodeContext.response = responseData.value;
                }
                const responseContentType = response.headers['content-type'] ?? '';
                if (autoDetectResponseFormat) {
                    if (responseContentType.includes('application/json')) {
                        responseFormat = 'json';
                    }
                    else if (genericRequestFunctions_1.binaryContentTypes.some((e) => responseContentType.includes(e))) {
                        responseFormat = 'file';
                    }
                    else {
                        responseFormat = 'text';
                    }
                }
                if (autoDetectResponseFormat && !fullResponse) {
                    delete response.headers;
                    delete response.statusCode;
                    delete response.statusMessage;
                }
                if (responseFormat === 'file') {
                    const outputPropertyName = this.getNodeParameter('options.response.response.outputPropertyName', 0, 'data');
                    const newItem = {
                        json: {},
                        binary: {},
                        pairedItem: {
                            item: itemIndex,
                        },
                    };
                    if (items[itemIndex].binary !== undefined) {
                        Object.assign(newItem.binary, items[itemIndex].binary);
                    }
                    let binaryData;
                    if (fullResponse) {
                        const returnItem = {};
                        for (const property of fullResponseProperties) {
                            if (property === 'body') {
                                continue;
                            }
                            returnItem[property] = response[property];
                        }
                        newItem.json = returnItem;
                        binaryData = response?.body;
                    }
                    else {
                        newItem.json = items[itemIndex].json;
                        binaryData = response;
                    }
                    const preparedBinaryData = await this.helpers.prepareBinaryData(binaryData, undefined, (0, parse_1.mimeTypeFromResponse)(responseContentType));
                    preparedBinaryData.fileName = (0, binaryData_1.setFilename)(preparedBinaryData, requestOptions, responseFileName);
                    newItem.binary[outputPropertyName] = preparedBinaryData;
                    returnItems.push(newItem);
                }
                else if (responseFormat === 'text') {
                    const outputPropertyName = this.getNodeParameter('options.response.response.outputPropertyName', 0, 'data');
                    if (fullResponse) {
                        const returnItem = {};
                        for (const property of fullResponseProperties) {
                            if (property === 'body') {
                                returnItem[outputPropertyName] = (0, genericRequestFunctions_1.toText)(response[property]);
                                continue;
                            }
                            returnItem[property] = response[property];
                        }
                        returnItems.push({
                            json: returnItem,
                            pairedItem: {
                                item: itemIndex,
                            },
                        });
                    }
                    else {
                        returnItems.push({
                            json: {
                                [outputPropertyName]: (0, genericRequestFunctions_1.toText)(response),
                            },
                            pairedItem: {
                                item: itemIndex,
                            },
                        });
                    }
                }
                else {
                    if (fullResponse) {
                        const returnItem = {};
                        for (const property of fullResponseProperties) {
                            returnItem[property] = response[property];
                        }
                        returnItems.push({
                            json: returnItem,
                            pairedItem: {
                                item: itemIndex,
                            },
                        });
                    }
                    else {
                        if (Array.isArray(response)) {
                            response.forEach((item) => returnItems.push({
                                json: item,
                                pairedItem: {
                                    item: itemIndex,
                                },
                            }));
                        }
                        else {
                            returnItems.push({
                                json: response,
                                pairedItem: {
                                    item: itemIndex,
                                },
                            });
                        }
                    }
                }
            }
        }
        catch (error) {
            if (!this.continueOnFail())
                throw error;
            returnItems.push({
                json: { error: error.message },
                pairedItem: { item: itemIndex },
            });
            continue;
        }
    }
    returnItems = returnItems.map(genericRequestFunctions_1.replaceNullValues);
    if (returnItems.length === 1 &&
        returnItems[0].json.data &&
        Array.isArray(returnItems[0].json.data)) {
        const message = 'To split the contents of ‘data’ into separate items for easier processing, add a ‘Split Out’ node after this one';
        if (this.addExecutionHints) {
            this.addExecutionHints({
                message,
                location: 'outputPane',
            });
        }
        else {
            this.logger.info(message);
        }
    }
    return [returnItems];
}
async function generateX402PaymentHeader(response, headers = {}) {
    let x402Config;
    const paymentRequiredHeader = headers['payment-required'] ?? headers['Payment-Required'];
    if (paymentRequiredHeader != null) {
        x402Config = JSON.parse(Buffer.from(paymentRequiredHeader, 'base64').toString('utf-8'));
    }
    else {
        x402Config = response;
    }
    if (x402Config.x402Version === 2) {
        const paymentConfig = x402Config.accepts[0];
        if (!paymentConfig) {
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'No payment config found');
        }
        if (paymentConfig.scheme !== 'exact') {
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Only exact scheme is supported for now');
        }
        this.logger.info(`x402 V2 Payment Requested, ${paymentConfig.amount} of token ${paymentConfig.asset} requested`);
        const { signature, data } = await (0, Wallets_1.getSignatureOperation)(this, 0, paymentConfig.payTo, paymentConfig.asset, paymentConfig.amount);
        const authorization = JSON.parse(data);
        authorization.validAfter = authorization.validAfter.toString();
        authorization.validBefore = authorization.validBefore.toString();
        const xPaymentObject = {
            x402Version: 2,
            accepted: paymentConfig,
            payload: {
                authorization,
                signature,
            },
            resource: x402Config.resource,
        };
        const jsonString = JSON.stringify(xPaymentObject);
        const base64Encoded = Buffer.from(jsonString, 'utf-8').toString('base64');
        return { paymentHeader: base64Encoded, headerKey: 'Payment-Signature' };
    }
    const paymentConfig = x402Config.accepts[0];
    if (!paymentConfig) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'No payment config found');
    }
    if (paymentConfig.scheme !== 'exact') {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Only exact scheme is supported for now');
    }
    this.logger.info(`x402 V1 Payment Requested, ${paymentConfig.maxAmountRequired} of token ${paymentConfig.asset} requested`);
    const { signature, data } = await (0, Wallets_1.getSignatureOperation)(this, 0, paymentConfig.payTo, paymentConfig.asset, paymentConfig.maxAmountRequired);
    const authorization = JSON.parse(data);
    authorization.validAfter = authorization.validAfter.toString();
    authorization.validBefore = authorization.validBefore.toString();
    const xPaymentObject = {
        x402Version: 1,
        scheme: 'exact',
        network: paymentConfig.network,
        payload: {
            authorization,
            signature,
        },
    };
    const jsonString = JSON.stringify(xPaymentObject);
    const base64Encoded = Buffer.from(jsonString, 'utf-8').toString('base64');
    return { paymentHeader: base64Encoded, headerKey: 'x-payment' };
}
