// n8n-nodes-1shot/dist/nodes/OneShot/executions/Chains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listChainsOperation = listChainsOperation;
exports.listChains = listChains;
const _1shot_1 = require("../types/1shot");
const constants_1 = require("../types/constants");
async function listChainsOperation(context, index) {
    const page = context.getNodeParameter('page', index);
    const pageSize = context.getNodeParameter('pageSize', index);
    return await listChains(context, page, pageSize);
}
async function listChains(context, page, pageSize) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'GET',
            url: `/chains`,
            qs: {
                pageSize: pageSize ?? 25,
                page: page ?? 1,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error listing supported Chains ${error.message}`, { error });
    }
    return new _1shot_1.PagedResponse([], 1, 1, 0);
}
