// n8n-nodes-1shot/dist/nodes/OneShot/executions/ContractEvents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listContractEventsOperation = listContractEventsOperation;
exports.getContractEventOperation = getContractEventOperation;
exports.createContractEventOperation = createContractEventOperation;
exports.updateContractEventOperation = updateContractEventOperation;
exports.deleteContractEventOperation = deleteContractEventOperation;
exports.searchContractEventOperation = searchContractEventOperation;
exports.listContractEvents = listContractEvents;
exports.createContractEvent = createContractEvent;
exports.getContractEvent = getContractEvent;
exports.updateContractEvent = updateContractEvent;
exports.deleteContractEvent = deleteContractEvent;
exports.searchContractEvent = searchContractEvent;
const n8n_workflow_1 = require("n8n-workflow");
const _1shot_1 = require("../types/1shot");
const constants_1 = require("../types/constants");
async function listContractEventsOperation(context, index) {
    const chainId = context.getNodeParameter('chainId', index);
    const page = context.getNodeParameter('page', index);
    const pageSize = context.getNodeParameter('pageSize', index);
    const name = context.getNodeParameter('name', index);
    const contractAddress = context.getNodeParameter('contractAddress', index);
    const eventName = context.getNodeParameter('eventName', index);
    return await listContractEvents(context, chainId == 'all' ? undefined : Number(chainId), page || undefined, pageSize || undefined, name || undefined, undefined, contractAddress || undefined, eventName || undefined);
}
async function getContractEventOperation(context, index) {
    const contractEventId = context.getNodeParameter('contractEventId', index);
    return await getContractEvent(context, contractEventId);
}
async function createContractEventOperation(context, index) {
    const chainId = context.getNodeParameter('chainId', index);
    const contractAddress = context.getNodeParameter('contractAddress', index);
    const name = context.getNodeParameter('name', index);
    const description = context.getNodeParameter('description', index);
    const eventName = context.getNodeParameter('eventName', index);
    return await createContractEvent(context, chainId, contractAddress, name, description, eventName);
}
async function updateContractEventOperation(context, index) {
    const contractEventId = context.getNodeParameter('contractEventId', index);
    const name = context.getNodeParameter('name', index);
    const description = context.getNodeParameter('description', index);
    return await updateContractEvent(context, contractEventId, name, description);
}
async function deleteContractEventOperation(context, index) {
    const contractEventId = context.getNodeParameter('contractEventId', index);
    return await deleteContractEvent(context, contractEventId);
}
async function searchContractEventOperation(context, index) {
    const contractEventId = context.getNodeParameter('contractEventId', index);
    const startBlock = context.getNodeParameter('startBlock', index);
    const endBlock = context.getNodeParameter('endBlock', index);
    const topicsData = context.getNodeParameter('topics', index);
    let topics = {};
    if (topicsData && topicsData.topic) {
        topics = {};
        for (const topic of topicsData.topic) {
            if (topic.name && topic.value) {
                topics[topic.name] = topic.value;
            }
        }
    }
    return await searchContractEvent(context, contractEventId, startBlock, endBlock, topics);
}
async function listContractEvents(context, chainId, page, pageSize, name, status, contractAddress, eventName) {
    try {
        const credentials = await context.getCredentials('oneShotOAuth2Api');
        const businessId = credentials.businessId;
        if (!businessId) {
            throw new n8n_workflow_1.NodeOperationError(context.getNode(), 'Business ID is required in credentials');
        }
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'GET',
            url: `/business/${businessId}/events`,
            qs: {
                pageSize: pageSize ?? 25,
                page: page ?? 1,
                chainId,
                name,
                status,
                contractAddress,
                eventName,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error listing Contract Events ${error.message}`, { error });
    }
    return new _1shot_1.PagedResponse([], 1, 1, 0);
}
async function createContractEvent(context, chainId, contractAddress, name, description, eventName) {
    try {
        const credentials = await context.getCredentials('oneShotOAuth2Api');
        const businessId = credentials.businessId;
        if (!businessId) {
            throw new n8n_workflow_1.NodeOperationError(context.getNode(), 'Business ID is required in credentials');
        }
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/business/${businessId}/events`,
            body: {
                chainId,
                contractAddress,
                name,
                description,
                eventName,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error creating Contract Event ${error.message}`, { error });
        throw error;
    }
}
async function getContractEvent(context, contractEventId) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'GET',
            url: `/events/${contractEventId}`,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error getting Contract Event ${error.message}`, { error });
        throw error;
    }
}
async function updateContractEvent(context, contractEventId, name, description) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'PUT',
            url: `/events/${contractEventId}`,
            body: {
                name,
                description,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error updating Contract Event ${error.message}`, { error });
        throw error;
    }
}
async function deleteContractEvent(context, contractEventId) {
    try {
        await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'DELETE',
            url: `/events/${contractEventId}`,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
    }
    catch (error) {
        context.logger.error(`Error deleting Contract Event ${error.message}`, { error });
        throw error;
    }
}
async function searchContractEvent(context, contractEventId, startBlock, endBlock, topics) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/events/${contractEventId}/search`,
            body: {
                startBlock,
                endBlock,
                topics,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error searching Contract Event ${error.message}`, { error });
        throw error;
    }
}
