// n8n-nodes-1shot/dist/nodes/OneShot/executions/ContractMethods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listContractMethodsOperation = listContractMethodsOperation;
exports.getContractMethodOperation = getContractMethodOperation;
exports.estimateContractMethodOperation = estimateContractMethodOperation;
exports.simulateContractMethodOperation = simulateContractMethodOperation;
exports.assureContractMethodsFromPromptOperation = assureContractMethodsFromPromptOperation;
exports.executeContractMethodOperation = executeContractMethodOperation;
exports.executeAsDelegatorContractMethodOperation = executeAsDelegatorContractMethodOperation;
exports.executeAndWaitContractMethodOperation = executeAndWaitContractMethodOperation;
exports.executeAsDelegatorAndWaitContractMethodOperation = executeAsDelegatorAndWaitContractMethodOperation;
exports.encodeContractMethodOperation = encodeContractMethodOperation;
exports.readContractMethodOperation = readContractMethodOperation;
exports.executeBatchOperation = executeBatchOperation;
exports.executeAsDelegatorBatchOperation = executeAsDelegatorBatchOperation;
exports.executeBatchAndWaitOperation = executeBatchAndWaitOperation;
exports.executeAsDelegatorBatchAndWaitOperation = executeAsDelegatorBatchAndWaitOperation;
exports.listContractMethods = listContractMethods;
exports.createContractMethod = createContractMethod;
exports.createContractMethodsFromAbi = createContractMethodsFromAbi;
exports.assureContractMethodsFromPrompt = assureContractMethodsFromPrompt;
exports.getContractMethod = getContractMethod;
exports.updateContractMethod = updateContractMethod;
exports.deleteContractMethod = deleteContractMethod;
exports.simulateContractMethod = simulateContractMethod;
exports.estimateContractMethod = estimateContractMethod;
exports.executeContractMethod = executeContractMethod;
exports.executeContractMethodAsDelegator = executeContractMethodAsDelegator;
exports.encodeContractMethod = encodeContractMethod;
exports.readContractMethod = readContractMethod;
exports.executeBatch = executeBatch;
exports.executeAsDelegatorBatch = executeAsDelegatorBatch;
const n8n_workflow_1 = require("n8n-workflow");
const _1shot_1 = require("../types/1shot");
const constants_1 = require("../types/constants");
const Transactions_1 = require("./Transactions");
async function listContractMethodsOperation(context, index) {
    const chainId = context.getNodeParameter('chainId', index);
    const page = context.getNodeParameter('page', index);
    const pageSize = context.getNodeParameter('pageSize', index);
    const name = context.getNodeParameter('name', index);
    const contractAddress = context.getNodeParameter('contractAddress', index);
    const promptId = context.getNodeParameter('promptId', index);
    const methodType = context.getNodeParameter('methodType', index);
    return await listContractMethods(context, chainId == 'all' ? undefined : Number(chainId), page || undefined, pageSize || undefined, name || undefined, undefined, contractAddress || undefined, promptId || undefined, methodType || undefined);
}
async function getContractMethodOperation(context, index) {
    const contractMethodId = context.getNodeParameter('contractMethodId', index);
    return await getContractMethod(context, contractMethodId);
}
async function estimateContractMethodOperation(context, index) {
    const contractMethodId = context.getNodeParameter('contractMethodId', index);
    const paramsString = context.getNodeParameter('params', index);
    const parsedParams = JSON.parse(paramsString);
    return await estimateContractMethod(context, contractMethodId, parsedParams);
}
async function simulateContractMethodOperation(context, index) {
    const contractMethodId = context.getNodeParameter('contractMethodId', index);
    const paramsString = context.getNodeParameter('params', index);
    const parsedParams = JSON.parse(paramsString);
    const gasLimit = context.getNodeParameter('gasLimit', index);
    return await simulateContractMethod(context, contractMethodId, parsedParams, gasLimit);
}
async function assureContractMethodsFromPromptOperation(context, index) {
    const chainId = context.getNodeParameter('chainId', index);
    const contractAddress = context.getNodeParameter('contractAddress', index);
    const walletId = context.getNodeParameter('walletId', index);
    const promptId = context.getNodeParameter('promptId', index);
    return await assureContractMethodsFromPrompt(context, chainId, contractAddress, walletId, promptId || undefined);
}
async function executeContractMethodOperation(context, index) {
    const contractMethodId = context.getNodeParameter('contractMethodId', index);
    const paramsString = context.getNodeParameter('params', index);
    const parsedParams = JSON.parse(paramsString);
    const additionalFields = context.getNodeParameter('additionalFields', index);
    const memo = additionalFields.memo;
    const walletId = additionalFields.walletId;
    const value = additionalFields.value;
    const gasLimit = additionalFields.gasLimit;
    const contractAddress = additionalFields.contractAddress;
    let authorizationList;
    if (additionalFields.authorizationList) {
        authorizationList = JSON.parse(additionalFields.authorizationList);
    }
    return await executeContractMethod(context, contractMethodId, parsedParams, walletId, memo, authorizationList, value, gasLimit, contractAddress);
}
async function executeAsDelegatorContractMethodOperation(context, index) {
    const contractMethodId = context.getNodeParameter('contractMethodId', index);
    const paramsString = context.getNodeParameter('params', index);
    const delegatorAddress = context.getNodeParameter('delegatorWalletAddress', index);
    const delegationId = context.getNodeParameter('delegationId', index);
    const delegationData = context.getNodeParameter('delegationData', index);
    const parsedParams = JSON.parse(paramsString);
    const additionalFields = context.getNodeParameter('additionalFields', index);
    const memo = additionalFields.memo;
    const walletId = additionalFields.walletId;
    const value = additionalFields.value;
    const gasLimit = additionalFields.gasLimit;
    const contractAddress = additionalFields.contractAddress;
    let authorizationList;
    if (additionalFields.authorizationList) {
        authorizationList = JSON.parse(additionalFields.authorizationList);
    }
    return await executeContractMethodAsDelegator(context, contractMethodId, parsedParams, delegatorAddress, delegationId, delegationData, walletId, memo, authorizationList, value, gasLimit, contractAddress);
}
async function executeAndWaitContractMethodOperation(context, index) {
    const contractMethodId = context.getNodeParameter('contractMethodId', index);
    const paramsString = context.getNodeParameter('params', index);
    const parsedParams = JSON.parse(paramsString);
    const additionalFields = context.getNodeParameter('additionalFields', index);
    const memo = additionalFields.memo;
    const walletId = additionalFields.walletId;
    const value = additionalFields.value;
    const gasLimit = additionalFields.gasLimit;
    const contractAddress = additionalFields.contractAddress;
    let authorizationList;
    if (additionalFields.authorizationList) {
        authorizationList = JSON.parse(additionalFields.authorizationList);
    }
    let transaction = await executeContractMethod(context, contractMethodId, parsedParams, walletId, memo, authorizationList, value, gasLimit, contractAddress);
    let status = 'Pending';
    let attempts = 0;
    while (status != 'Completed' && status != 'Failed' && attempts < 20) {
        await (0, n8n_workflow_1.sleep)(2000);
        transaction = await (0, Transactions_1.getTransaction)(context, transaction.id);
        status = transaction.status;
        attempts++;
    }
    if (status === 'Failed') {
        return { success: false, result: transaction };
    }
    return { success: true, result: transaction };
}
async function executeAsDelegatorAndWaitContractMethodOperation(context, index) {
    const contractMethodId = context.getNodeParameter('contractMethodId', index);
    const paramsString = context.getNodeParameter('params', index);
    const delegatorAddress = context.getNodeParameter('delegatorWalletAddress', index);
    const parsedParams = JSON.parse(paramsString);
    const additionalFields = context.getNodeParameter('additionalFields', index);
    const memo = additionalFields.memo;
    const walletId = additionalFields.walletId;
    const value = additionalFields.value;
    const gasLimit = additionalFields.gasLimit;
    const contractAddress = additionalFields.contractAddress;
    const delegationId = additionalFields.delegationId;
    const delegationData = additionalFields.delegationData;
    let authorizationList;
    if (additionalFields.authorizationList) {
        authorizationList = JSON.parse(additionalFields.authorizationList);
    }
    let transaction = await executeContractMethodAsDelegator(context, contractMethodId, parsedParams, delegatorAddress, delegationId, delegationData, walletId, memo, authorizationList, value, gasLimit, contractAddress);
    let status = 'Pending';
    while (status != 'Completed' && status != 'Failed') {
        await (0, n8n_workflow_1.sleep)(2000);
        transaction = await (0, Transactions_1.getTransaction)(context, transaction.id);
        status = transaction.status;
    }
    if (status === 'Failed') {
        return { success: false, result: transaction };
    }
    return { success: true, result: transaction };
}
async function encodeContractMethodOperation(context, index) {
    const contractMethodId = context.getNodeParameter('contractMethodId', index);
    const paramsString = context.getNodeParameter('params', index);
    const parsedParams = JSON.parse(paramsString);
    const additionalFields = context.getNodeParameter('additionalFields', index);
    const value = additionalFields.value;
    return await encodeContractMethod(context, contractMethodId, parsedParams, value);
}
async function readContractMethodOperation(context, index) {
    const contractMethodId = context.getNodeParameter('contractMethodId', index);
    const paramsString = context.getNodeParameter('params', index);
    const parsedParams = JSON.parse(paramsString);
    return await readContractMethod(context, contractMethodId, parsedParams);
}
async function executeBatchOperation(context, index) {
    const contractMethodsData = context.getNodeParameter('contractMethods', index);
    const walletId = context.getNodeParameter('walletId', index);
    const additionalFields = context.getNodeParameter('additionalFields', index);
    const contractMethods = contractMethodsData.contractMethod.map((method, methodIndex) => {
        const parsedParams = JSON.parse(method.params);
        const optionalFields = method.optionalFields || {};
        return {
            contractMethodId: method.contractMethodId,
            executionIndex: methodIndex,
            params: parsedParams,
            contractAddress: optionalFields.contractAddress || undefined,
            value: optionalFields.value || undefined,
        };
    });
    let authorizationList;
    if (additionalFields.authorizationList) {
        authorizationList = JSON.parse(additionalFields.authorizationList);
    }
    return await executeBatch(context, contractMethods, walletId, additionalFields.atomic, additionalFields.memo, authorizationList, additionalFields.gasLimit);
}
async function executeAsDelegatorBatchOperation(context, index) {
    const contractMethodsData = context.getNodeParameter('contractMethods', index);
    const walletId = context.getNodeParameter('walletId', index);
    const additionalFields = context.getNodeParameter('additionalFields', index);
    const contractMethods = contractMethodsData.contractMethod.map((method, methodIndex) => {
        const parsedParams = JSON.parse(method.params);
        const optionalFields = method.optionalFields || {};
        return {
            contractMethodId: method.contractMethodId,
            executionIndex: methodIndex,
            params: parsedParams,
            delegatorAddress: method.delegatorAddress,
            delegationId: method.delegationId,
            delegationData: method.delegationData,
            contractAddress: optionalFields.contractAddress || undefined,
            value: optionalFields.value || undefined,
        };
    });
    let authorizationList;
    if (additionalFields.authorizationList) {
        authorizationList = JSON.parse(additionalFields.authorizationList);
    }
    return await executeAsDelegatorBatch(context, contractMethods, walletId, additionalFields.atomic, additionalFields.memo, authorizationList, additionalFields.gasLimit);
}
async function executeBatchAndWaitOperation(context, index) {
    const contractMethodsData = context.getNodeParameter('contractMethods', index);
    const walletId = context.getNodeParameter('walletId', index);
    const additionalFields = context.getNodeParameter('additionalFields', index);
    const contractMethods = contractMethodsData.contractMethod.map((method, methodIndex) => {
        const parsedParams = JSON.parse(method.params);
        const optionalFields = method.optionalFields || {};
        return {
            contractMethodId: method.contractMethodId,
            executionIndex: methodIndex,
            params: parsedParams,
            contractAddress: optionalFields.contractAddress || undefined,
            value: optionalFields.value || undefined,
        };
    });
    let authorizationList;
    if (additionalFields.authorizationList) {
        authorizationList = JSON.parse(additionalFields.authorizationList);
    }
    let transaction = await executeBatch(context, contractMethods, walletId, additionalFields.atomic, additionalFields.memo, authorizationList, additionalFields.gasLimit);
    let status = 'Pending';
    let attempts = 0;
    while (status != 'Completed' && status != 'Failed' && attempts < 20) {
        await (0, n8n_workflow_1.sleep)(2000);
        transaction = await (0, Transactions_1.getTransaction)(context, transaction.id);
        status = transaction.status;
        attempts++;
    }
    if (status === 'Failed') {
        return { success: false, result: transaction };
    }
    return { success: true, result: transaction };
}
async function executeAsDelegatorBatchAndWaitOperation(context, index) {
    const contractMethodsData = context.getNodeParameter('contractMethods', index);
    const walletId = context.getNodeParameter('walletId', index);
    const additionalFields = context.getNodeParameter('additionalFields', index);
    const contractMethods = contractMethodsData.contractMethod.map((method, methodIndex) => {
        const parsedParams = JSON.parse(method.params);
        const optionalFields = method.optionalFields || {};
        return {
            contractMethodId: method.contractMethodId,
            executionIndex: methodIndex,
            params: parsedParams,
            delegatorAddress: method.delegatorAddress,
            contractAddress: optionalFields.contractAddress || undefined,
            value: optionalFields.value || undefined,
        };
    });
    let authorizationList;
    if (additionalFields.authorizationList) {
        authorizationList = JSON.parse(additionalFields.authorizationList);
    }
    let transaction = await executeAsDelegatorBatch(context, contractMethods, walletId, additionalFields.atomic, additionalFields.memo, authorizationList, additionalFields.gasLimit);
    let status = 'Pending';
    let attempts = 0;
    while (status != 'Completed' && status != 'Failed' && attempts < 20) {
        await (0, n8n_workflow_1.sleep)(2000);
        transaction = await (0, Transactions_1.getTransaction)(context, transaction.id);
        status = transaction.status;
        attempts++;
    }
    if (status === 'Failed') {
        return { success: false, result: transaction };
    }
    return { success: true, result: transaction };
}
async function listContractMethods(context, chainId, page, pageSize, name, status, contractAddress, promptId, methodType) {
    try {
        const credentials = await context.getCredentials('oneShotOAuth2Api');
        const businessId = credentials.businessId;
        if (!businessId) {
            throw new n8n_workflow_1.NodeOperationError(context.getNode(), 'Business ID is required in credentials');
        }
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'GET',
            url: `/business/${businessId}/methods`,
            qs: {
                pageSize: pageSize ?? 25,
                page: page ?? 1,
                chainId,
                name,
                status,
                contractAddress,
                promptId,
                methodType,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error listing Contract Methods ${error.message}`, { error });
    }
    return new _1shot_1.PagedResponse([], 1, 1, 0);
}
async function createContractMethod(context, chainId, contractAddress, walletId, name, description, functionName, stateMutability, inputs, outputs, callbackUrl) {
    try {
        const credentials = await context.getCredentials('oneShotOAuth2Api');
        const businessId = credentials.businessId;
        if (!businessId) {
            throw new n8n_workflow_1.NodeOperationError(context.getNode(), 'Business ID is required in credentials');
        }
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/business/${businessId}/methods`,
            body: {
                chainId,
                contractAddress,
                walletId,
                name,
                description,
                functionName,
                stateMutability,
                inputs,
                outputs,
                callbackUrl,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error creating Contract Method ${error.message}`, { error });
        throw error;
    }
}
async function createContractMethodsFromAbi(context, chainId, contractAddress, walletId, abi, name, description, tags) {
    try {
        const credentials = await context.getCredentials('oneShotOAuth2Api');
        const businessId = credentials.businessId;
        if (!businessId) {
            throw new n8n_workflow_1.NodeOperationError(context.getNode(), 'Business ID is required in credentials');
        }
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/business/${businessId}/methods/abi`,
            body: {
                chainId,
                contractAddress,
                walletId,
                abi,
                name,
                description,
                tags,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error creating Contract Methods from ABI ${error.message}`, { error });
        throw error;
    }
}
async function assureContractMethodsFromPrompt(context, chainId, contractAddress, walletId, promptId) {
    try {
        const credentials = await context.getCredentials('oneShotOAuth2Api');
        const businessId = credentials.businessId;
        if (!businessId) {
            throw new n8n_workflow_1.NodeOperationError(context.getNode(), 'Business ID is required in credentials');
        }
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/business/${businessId}/methods/prompt`,
            body: {
                chainId,
                contractAddress,
                walletId,
                promptId,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error assuring Contract Methods from Prompt ${error.message}`, { error });
        throw error;
    }
}
async function getContractMethod(context, contractMethodId) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'GET',
            url: `/methods/${contractMethodId}`,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error getting Contract Method ${error.message}`, { error });
        throw error;
    }
}
async function updateContractMethod(context, contractMethodId, chainId, contractAddress, walletId, name, description, functionName, payable, nativeTransaction, callbackUrl) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'PUT',
            url: `/methods/${contractMethodId}`,
            body: {
                chainId,
                contractAddress,
                walletId,
                name,
                description,
                functionName,
                payable,
                nativeTransaction,
                callbackUrl,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error updating Contract Method ${error.message}`, { error });
        throw error;
    }
}
async function deleteContractMethod(context, contractMethodId) {
    try {
        await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'DELETE',
            url: `/methods/${contractMethodId}`,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
    }
    catch (error) {
        context.logger.error(`Error deleting Contract Method ${error.message}`, { error });
        throw error;
    }
}
async function simulateContractMethod(context, contractMethodId, params, gasLimit) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/methods/${contractMethodId}/test`,
            body: {
                params,
                gasLimit,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error testing Contract Method ${error.message}`, { error });
        throw error;
    }
}
async function estimateContractMethod(context, contractMethodId, params) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/methods/${contractMethodId}/estimate`,
            body: {
                params,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error estimating Contract Method ${error.message}`, { error });
        throw error;
    }
}
async function executeContractMethod(context, contractMethodId, params, walletId, memo, authorizationList, value, gasLimit, contractAddress) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/methods/${contractMethodId}/execute`,
            body: {
                params,
                walletId,
                memo,
                authorizationList,
                value,
                gasLimit,
                contractAddress,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error executing Contract Method ${error.message}`, { error });
        throw error;
    }
}
async function executeContractMethodAsDelegator(context, contractMethodId, params, delegatorAddress, delegationId, delegationData, walletId, memo, authorizationList, value, gasLimit, contractAddress) {
    try {
        const body = {
            params,
            walletId,
            memo,
            authorizationList,
            value,
            gasLimit,
            contractAddress,
        };
        if (delegationId) {
            body.delegationId = delegationId;
        }
        else if (delegationData) {
            body.delegationData = delegationData;
        }
        else if (delegatorAddress) {
            body.delegatorAddress = delegatorAddress;
        }
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/methods/${contractMethodId}/executeAsDelegator`,
            body,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error executing Contract Method ${error.message}`, { error });
        throw error;
    }
}
async function encodeContractMethod(context, contractMethodId, params, value, authorizationList) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/methods/${contractMethodId}/encode`,
            body: { params, value, authorizationList },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error encoding Contract Method ${error.message}`, { error });
        throw error;
    }
}
async function readContractMethod(context, contractMethodId, params) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/methods/${contractMethodId}/read`,
            body: { params },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error reading Contract Method ${error.message}`, { error });
        throw error;
    }
}
async function executeBatch(context, contractMethods, walletId, atomic, memo, authorizationList, gasLimit) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/methods/executeBatch`,
            body: {
                contractMethods,
                walletId,
                atomic,
                memo,
                authorizationList,
                gasLimit,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error executing Contract Method batch ${error.message}`, { error });
        throw error;
    }
}
async function executeAsDelegatorBatch(context, contractMethods, walletId, atomic, memo, authorizationList, gasLimit) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/methods/executeAsDelegatorBatch`,
            body: {
                contractMethods,
                walletId,
                atomic,
                memo,
                authorizationList,
                gasLimit,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error executing Contract Method batch as delegator ${error.message}`, {
            error,
        });
        throw error;
    }
}
