// n8n-nodes-1shot/dist/nodes/OneShot/executions/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadChainOptions = loadChainOptions;
exports.loadChainOptionsWithAll = loadChainOptionsWithAll;
exports.loadContractMethodExecutionOptions = loadContractMethodExecutionOptions;
exports.loadContractMethodReadOptions = loadContractMethodReadOptions;
exports.loadContractMethodAllOptions = loadContractMethodAllOptions;
exports.loadContractEventOptions = loadContractEventOptions;
exports.loadX402TokenOptions = loadX402TokenOptions;
const ContractMethods_1 = require("./ContractMethods");
const ContractEvents_1 = require("./ContractEvents");
const Chains_1 = require("./Chains");
const x402_1 = require("./x402");
async function loadChainOptions() {
    const options = [];
    const chains = await (0, Chains_1.listChains)(this, 1, 1000);
    for (const chain of chains.response) {
        options.push({
            name: chain.name,
            value: chain.chainId,
            description: `${chain.name} (${chain.chainId}-${chain.type})`,
        });
    }
    return options;
}
async function loadChainOptionsWithAll() {
    const options = [];
    const chains = await (0, Chains_1.listChains)(this, 1, 1000);
    options.push({
        name: 'All',
        value: 'all',
        description: 'All chains',
    });
    for (const chain of chains.response) {
        options.push({
            name: chain.name,
            value: chain.chainId,
            description: `${chain.name} (${chain.chainId}-${chain.type})`,
        });
    }
    return options;
}
async function loadContractMethodExecutionOptions() {
    return loadContractMethodOptions(this, 'write');
}
async function loadContractMethodReadOptions() {
    return loadContractMethodOptions(this, 'read');
}
async function loadContractMethodAllOptions() {
    return loadContractMethodOptions(this, undefined);
}
async function loadContractMethodOptions(loadOptionsFunctions, methodType) {
    const options = [];
    const contractMethods = await (0, ContractMethods_1.listContractMethods)(loadOptionsFunctions, undefined, 1, 1000, undefined, undefined, undefined, undefined, methodType);
    for (const contractMethod of contractMethods.response) {
        options.push({
            name: `${contractMethod.name} (${contractMethod.chainId})`,
            value: contractMethod.id,
            description: contractMethod.description || '',
        });
    }
    return options;
}
async function loadContractEventOptions() {
    const options = [];
    const contractEvents = await (0, ContractEvents_1.listContractEvents)(this, undefined, 1, 1000, undefined, undefined, undefined, undefined);
    for (const contractEvent of contractEvents.response) {
        options.push({
            name: `${contractEvent.name} (${contractEvent.chainId})`,
            value: contractEvent.id,
            description: contractEvent.description || '',
        });
    }
    return options;
}
async function loadX402TokenOptions() {
    const options = [];
    const supportedPayments = await (0, x402_1.getX402Supported)(this, true);
    for (const supportedPayment of supportedPayments.kinds) {
        for (const token of supportedPayment.tokens) {
            options.push({
                name: `${token.name} (${supportedPayment.network})`,
                value: `${supportedPayment.network}:${token.contractAddress}`,
                description: `${token.name} (${token.contractAddress}) on ${supportedPayment.network}`,
            });
        }
    }
    return options;
}
