// n8n-nodes-1shot/dist/nodes/OneShot/executions/Prompts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.searchPromptsOperation = searchPromptsOperation;
exports.searchPrompts = searchPrompts;
const constants_1 = require("../types/constants");
async function searchPromptsOperation(context, index) {
    const query = context.getNodeParameter('query', index);
    const chainId = context.getNodeParameter('chainId', index);
    return await searchPrompts(context, query, chainId == 'all' ? undefined : Number(chainId));
}
async function searchPrompts(context, query, chainId) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: '/prompts/search',
            body: {
                query,
                chainId,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error searching prompts ${error.message}`, { error });
        throw error;
    }
}
