// n8n-nodes-1shot/dist/nodes/OneShot/executions/Structs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateStruct = updateStruct;
exports.addStructParam = addStructParam;
exports.updateStructParams = updateStructParams;
exports.deleteStructParam = deleteStructParam;
const constants_1 = require("../types/constants");
async function updateStruct(context, structId, name) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'PUT',
            url: `/structs/${structId}`,
            body: {
                name,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error updating struct ${error.message}`, { error });
        throw error;
    }
}
async function addStructParam(context, structId, param) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/structs/${structId}/params`,
            body: param,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error adding struct param ${error.message}`, { error });
        throw error;
    }
}
async function updateStructParams(context, structId, updates) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'PUT',
            url: `/structs/${structId}/params`,
            body: {
                updates,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error updating struct params ${error.message}`, { error });
        throw error;
    }
}
async function deleteStructParam(context, structId, structParamId) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'DELETE',
            url: `/structs/${structId}/params/${structParamId}`,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error deleting struct param ${error.message}`, { error });
        throw error;
    }
}
