// n8n-nodes-1shot/dist/nodes/OneShot/executions/Transactions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listTransactionsOperation = listTransactionsOperation;
exports.getTransactionOperation = getTransactionOperation;
exports.listTransactions = listTransactions;
exports.getTransaction = getTransaction;
const n8n_workflow_1 = require("n8n-workflow");
const constants_1 = require("../types/constants");
async function listTransactionsOperation(context, index) {
    const chainId = context.getNodeParameter('chainId', index);
    const page = context.getNodeParameter('page', index);
    const pageSize = context.getNodeParameter('pageSize', index);
    const status = context.getNodeParameter('status', index);
    const walletId = context.getNodeParameter('walletId', index);
    const contractMethodId = context.getNodeParameter('contractMethodId', index);
    const apiCredentialId = context.getNodeParameter('apiCredentialId', index);
    const userId = context.getNodeParameter('userId', index);
    const memo = context.getNodeParameter('memo', index);
    const createdAfter = context.getNodeParameter('createdAfter', index);
    const createdBefore = context.getNodeParameter('createdBefore', index);
    return await listTransactions(context, chainId == 'all' ? undefined : Number(chainId), page || undefined, pageSize || undefined, status == 'none' ? undefined : status, walletId || undefined, contractMethodId || undefined, apiCredentialId || undefined, userId || undefined, memo || undefined, createdAfter || undefined, createdBefore || undefined);
}
async function getTransactionOperation(context, index) {
    const transactionId = context.getNodeParameter('transactionId', index);
    return await getTransaction(context, transactionId);
}
async function listTransactions(context, chainId, page, pageSize, status, walletId, contractMethodId, apiCredentialId, userId, memo, createdAfter, createdBefore) {
    try {
        const credentials = await context.getCredentials('oneShotOAuth2Api');
        const businessId = credentials.businessId;
        if (!businessId) {
            throw new n8n_workflow_1.NodeOperationError(context.getNode(), 'Business ID is required in credentials');
        }
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'GET',
            url: `/business/${businessId}/transactions`,
            qs: {
                chainId,
                pageSize: pageSize ?? 25,
                page: page ?? 1,
                status,
                walletId,
                contractMethodId,
                apiCredentialId,
                userId,
                memo,
                createdAfter,
                createdBefore,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error listing transactions ${error.message}`, { error });
        throw error;
    }
}
async function getTransaction(context, transactionId) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'GET',
            url: `/transactions/${transactionId}`,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error getting transaction ${error.message}`, { error });
        throw error;
    }
}
