// n8n-nodes-1shot/dist/nodes/OneShot/executions/Wallets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listWalletsOperation = listWalletsOperation;
exports.createWalletOperation = createWalletOperation;
exports.getWalletOperation = getWalletOperation;
exports.updateWalletOperation = updateWalletOperation;
exports.deleteWalletOperation = deleteWalletOperation;
exports.getSignatureOperation = getSignatureOperation;
exports.loadWalletOptions = loadWalletOptions;
exports.listWallets = listWallets;
exports.createWallet = createWallet;
exports.getWallet = getWallet;
exports.updateWallet = updateWallet;
exports.deleteWallet = deleteWallet;
exports.getSignature = getSignature;
const n8n_workflow_1 = require("n8n-workflow");
const _1shot_1 = require("../types/1shot");
const constants_1 = require("../types/constants");
async function listWalletsOperation(context, index) {
    const chainId = context.getNodeParameter('chainId', index);
    const page = context.getNodeParameter('page', index);
    const pageSize = context.getNodeParameter('pageSize', index);
    const name = context.getNodeParameter('name', index);
    return await listWallets(context, chainId == 'all' ? undefined : Number(chainId), page || undefined, pageSize || undefined, name || undefined);
}
async function createWalletOperation(context, index) {
    const chainId = context.getNodeParameter('chainId', index);
    const name = context.getNodeParameter('name', index);
    const description = context.getNodeParameter('description', index);
    return await createWallet(context, chainId, name, description || undefined);
}
async function getWalletOperation(context, index) {
    const walletId = context.getNodeParameter('walletId', index);
    return await getWallet(context, walletId, true);
}
async function updateWalletOperation(context, index) {
    const walletId = context.getNodeParameter('walletId', index);
    const name = context.getNodeParameter('name', index);
    const description = context.getNodeParameter('description', index);
    return await updateWallet(context, walletId, name || undefined, description || undefined);
}
async function deleteWalletOperation(context, index) {
    const walletId = context.getNodeParameter('walletId', index);
    return await deleteWallet(context, walletId);
}
async function getSignatureOperation(context, index, payTo, asset, amount) {
    const walletId = context.getNodeParameter('walletId', index);
    const type = 'erc3009';
    const now = Math.floor(Date.now() / 1000);
    const validUntil = now + 90;
    const validAfter = now - 90;
    const signatureResponse = await getSignature(context, walletId, type, asset, payTo, amount, validUntil, validAfter);
    return signatureResponse;
}
async function loadWalletOptions() {
    const response = await listWallets(this, undefined, 1, 1000, undefined);
    const options = [];
    for (const wallet of response.response) {
        options.push({
            name: wallet.name,
            value: wallet.id,
            description: wallet.description,
        });
    }
    return options;
}
async function listWallets(context, chainId, page, pageSize, name) {
    try {
        const credentials = await context.getCredentials('oneShotOAuth2Api');
        const businessId = credentials.businessId;
        if (!businessId) {
            throw new n8n_workflow_1.NodeOperationError(context.getNode(), 'Business ID is required in credentials');
        }
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'GET',
            url: `/business/${businessId}/wallets`,
            qs: {
                pageSize: pageSize ?? 25,
                page: page ?? 1,
                chainId: chainId ?? undefined,
                name: name ?? undefined,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error loading wallets ${error.message}`, { error });
    }
    return new _1shot_1.PagedResponse([], 1, 1, 0);
}
async function createWallet(context, chainId, name, description) {
    try {
        const credentials = await context.getCredentials('oneShotOAuth2Api');
        const businessId = credentials.businessId;
        if (!businessId) {
            throw new n8n_workflow_1.NodeOperationError(context.getNode(), 'Business ID is required in credentials');
        }
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: `/business/${businessId}/wallets`,
            body: {
                chainId,
                name,
                description,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error creating wallet ${error.message}`, { error });
        throw error;
    }
}
async function getWallet(context, walletId, includeBalances) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'GET',
            url: `/wallets/${walletId}`,
            qs: {
                includeBalances: includeBalances ?? false,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error getting wallet ${error.message}`, { error });
        throw error;
    }
}
async function updateWallet(context, walletId, name, description) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'PUT',
            url: `/wallets/${walletId}`,
            body: {
                name,
                description,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error updating wallet ${error.message}`, { error });
        throw error;
    }
}
async function deleteWallet(context, walletId) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'DELETE',
            url: `/wallets/${walletId}`,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error deleting wallet ${error.message}`, { error });
        throw error;
    }
}
async function getSignature(context, walletId, type, contractAddress, destinationAddress, amount, validUntil, validAfter) {
    try {
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'GET',
            url: `/wallets/${walletId}/signature/${type}`,
            qs: {
                contractAddress,
                destinationAddress,
                amount: amount,
                validUntil: validUntil,
                validAfter: validAfter ?? undefined,
            },
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error getting signature ${error.message}`, { error });
        throw error;
    }
}
