// n8n-nodes-1shot/dist/nodes/OneShot/executions/Webhooks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.webhookTrigger = webhookTrigger;
const ED25519_1 = require("../crypto/ED25519");
const x402_1 = require("./x402");
const webhookUtils_1 = require("../utils/webhookUtils");
let x402ScanRegistrationInProcess = false;
async function webhookTrigger() {
    const webhookType = this.getNodeParameter('webhookType');
    const body = this.getBodyData();
    if (webhookType === 'oneshot') {
        return await handleOneShotWebhook.call(this, body);
    }
    else if (webhookType === 'x402') {
        return await handleX402Webhook.call(this, body);
    }
    else {
        throw new Error(`Unsupported webhook type: ${webhookType}`);
    }
}
async function handleOneShotWebhook(body) {
    const publicKey = this.getNodeParameter('publicKey');
    const signature = body.signature;
    if (!signature) {
        throw new Error('No signature provided in webhook payload');
    }
    const { signature: _, ...payloadWithoutSignature } = body;
    const isValid = await verify1ShotSignature(publicKey, signature, payloadWithoutSignature);
    if (!isValid) {
        throw new Error('1Shot: Signature verification failed');
    }
    return {
        workflowData: [
            [
                {
                    json: body,
                },
            ],
        ],
    };
}
async function handleX402Webhook(_body) {
    const responseMode = this.getNodeParameter('responseMode', 'onReceived');
    const nodeStaticData = this.getWorkflowStaticData('node');
    const options = this.getNodeParameter('options', {});
    this.logger.debug('x402 webhook options', {
        responseMode,
        responseHeaders: options.responseHeaders,
        x402RefundsContactEmail: options.x402RefundsContactEmail ?? '',
    });
    const headers = this.getHeaderData();
    const req = this.getRequestObject();
    const resp = this.getResponseObject();
    const requestMethod = this.getRequestObject().method;
    if (!(0, webhookUtils_1.isIpWhitelisted)(options.ipWhitelist, req.ips, req.ip)) {
        resp.writeHead(403);
        resp.end('IP is not whitelisted to access the webhook!');
        return { noWebhookResponse: true };
    }
    const prepareOutput = (0, webhookUtils_1.setupOutputConnection)(this, requestMethod, {});
    if (options.rawBody && !req.rawBody) {
        await req.readRawBody();
    }
    const credentials = await this.getCredentials('oneShotOAuth2Api');
    if (!credentials) {
        resp.writeHead(403);
        resp.end('oneShotOAuth2Api credential not found');
        return { noWebhookResponse: true };
    }
    const supportedTokens = await (0, x402_1.getX402Supported)(this);
    let configuredTokens = this.getNodeParameter('tokens');
    if (configuredTokens == null) {
        configuredTokens = { paymentToken: [] };
    }
    const { resourceDescription, mimeType } = options;
    const responseData = this.getNodeParameter('responseData');
    const httpMethod = this.getNodeParameter('httpMethod');
    const webhookUrl = this.getNodeWebhookUrl('default');
    if (webhookUrl == null) {
        resp.writeHead(403);
        resp.end('webhookUrl not found');
        return { noWebhookResponse: true };
    }
    const shouldRegisterWithX402Scan = (nodeStaticData.x402ScanRegistered != webhookUrl ||
        nodeStaticData.resourceDescription != resourceDescription ||
        nodeStaticData.mimeType != mimeType) &&
        !x402ScanRegistrationInProcess &&
        (httpMethod === 'POST' || httpMethod === 'GET');
    if (shouldRegisterWithX402Scan) {
        try {
            x402ScanRegistrationInProcess = true;
            await registerWebhookWithX402Scan(this, webhookUrl);
            nodeStaticData.x402ScanRegistered = webhookUrl;
            nodeStaticData.resourceDescription = resourceDescription;
            nodeStaticData.mimeType = mimeType;
            x402ScanRegistrationInProcess = false;
            this.logger.info('Successfully registered workflow on x402Scan');
        }
        catch (err) {
            this.logger.error('Error registering node on x402Scan:', err);
            x402ScanRegistrationInProcess = false;
        }
    }
    const paymentRequirements = new Array();
    const configuredKinds = new Array();
    for (const configuredToken of configuredTokens.paymentToken) {
        const [network, contractAddress] = configuredToken.paymentToken.split(':');
        if (configuredKinds.includes(network)) {
            resp.writeHead(403);
            resp.end(`Misconfiguration: Network ${network} has multiple configured tokens. You may only have one payment token per network.`);
            return { noWebhookResponse: true };
        }
        configuredKinds.push(network);
        const kind = supportedTokens.kinds.find((kind) => kind.network === network);
        if (kind == null) {
            throw new Error(`Supported network ${network} not found`);
        }
        const supportedToken = kind.tokens.find((token) => token.contractAddress === contractAddress);
        if (supportedToken == null) {
            throw new Error(`Supported token ${contractAddress} not found`);
        }
        paymentRequirements.push(new PaymentRequirements(kind.scheme, kind.network, configuredToken.paymentAmount.toString(), webhookUrl, resourceDescription || 'OneShot API Webhook', mimeType || 'application/json', undefined, configuredToken.payToAddress, 60, supportedToken.contractAddress, {
            name: supportedToken.name,
            version: supportedToken.version,
        }));
    }
    const xPaymentHeader = headers['x-payment'];
    if (!xPaymentHeader == null || typeof xPaymentHeader !== 'string') {
        return generateX402Error(resp, 'No x-payment header provided', paymentRequirements, options.x402RefundsContactEmail);
    }
    try {
        const decodedXPayment = Buffer.from(xPaymentHeader, 'base64').toString('utf-8');
        const decodedXPaymentJson = JSON.parse(decodedXPayment);
        const validation = validateXPayment(decodedXPaymentJson);
        if (validation != 'valid') {
            return generateX402Error(resp, 'x-payment header is not valid', paymentRequirements, options.x402RefundsContactEmail);
        }
        const verification = verifyPaymentDetails(decodedXPaymentJson, paymentRequirements);
        if (!verification.valid) {
            return generateX402Error(resp, `x-payment header is not valid for reasons: ${verification.errors}`, paymentRequirements, options.x402RefundsContactEmail);
        }
        const verifyResponse = await (0, x402_1.verifyX402Payment)(this, decodedXPaymentJson.x402Version, decodedXPaymentJson, verification.paymentRequirements);
        if (!verifyResponse.isValid) {
            return generateX402Error(resp, `x-payment verification failed: ${verifyResponse.invalidReason}`, paymentRequirements, options.x402RefundsContactEmail);
        }
        try {
            const settleResponse = await (0, x402_1.settleX402Payment)(this, decodedXPaymentJson.x402Version, decodedXPaymentJson, verification.paymentRequirements);
            if (!settleResponse.success) {
                return generateX402Error(resp, `x-payment settlement failed: ${settleResponse.error}`, paymentRequirements, options.x402RefundsContactEmail);
            }
            return generateResponse(this, req, responseMode, responseData, settleResponse.txHash, paymentRequirements, decodedXPaymentJson, prepareOutput, options.x402RefundsContactEmail);
        }
        catch (error) {
            this.logger.error('Error in x402 webhook settlement, moving on...', error);
            return generateResponse(this, req, responseMode, responseData, 'TBD', paymentRequirements, decodedXPaymentJson, prepareOutput, options.x402RefundsContactEmail);
        }
    }
    catch (error) {
        this.logger.error('Error in x402 webhook', error);
        return generateX402Error(resp, `No x-payment header provided: ${error.message}`, paymentRequirements, options.x402RefundsContactEmail);
    }
}
function generateResponse(context, req, responseMode, responseData, txHash, paymentRequirements, paymentPayload, prepareOutput, x402RefundsContactEmail) {
    const response = {
        json: {
            headers: req.headers,
            params: req.params,
            query: req.query,
            body: req.body,
            txHash: txHash,
            paymentRequirements: paymentRequirements,
            paymentPayload: paymentPayload,
        },
    };
    if (responseMode === 'streaming') {
        const res = context.getResponseObject();
        const x402RefundsHeader = (0, webhookUtils_1.getX402RefundHeader)(x402RefundsContactEmail);
        if (x402RefundsHeader != null) {
            res.writeHead(200, {
                'Content-Type': 'application/json; charset=utf-8',
                'Transfer-Encoding': 'chunked',
                'Cache-Control': 'no-cache',
                Link: x402RefundsHeader,
                Connection: 'keep-alive',
            });
        }
        else {
            res.writeHead(200, {
                'Content-Type': 'application/json; charset=utf-8',
                'Transfer-Encoding': 'chunked',
                'Cache-Control': 'no-cache',
                Connection: 'keep-alive',
            });
        }
        res.flushHeaders();
        return {
            noWebhookResponse: true,
            workflowData: prepareOutput(response),
        };
    }
    return {
        webhookResponse: responseData,
        workflowData: prepareOutput(response),
    };
}
function generateX402Error(resp, errorMessage, paymentRequirements, x402RefundsContactEmail) {
    const x402RefundsHeader = (0, webhookUtils_1.getX402RefundHeader)(x402RefundsContactEmail);
    if (x402RefundsHeader != null) {
        resp.writeHead(402, {
            'Content-Type': 'application/json',
            Link: x402RefundsHeader,
        });
    }
    else {
        resp.writeHead(402, { 'Content-Type': 'application/json' });
    }
    resp.end(JSON.stringify({
        x402Version: 1,
        error: errorMessage,
        accepts: paymentRequirements,
    }));
    return { noWebhookResponse: true };
}
async function verify1ShotSignature(publicKey, signature, payload) {
    try {
        const publicKeyBytes = Buffer.from(publicKey, 'base64');
        const signatureBytes = Buffer.from(signature, 'base64');
        const sortedData = sortObjectKeys(payload);
        const message = JSON.stringify(sortedData);
        const messageBytes = new TextEncoder().encode(message);
        return await (0, ED25519_1.verifyAsync)(signatureBytes, messageBytes, publicKeyBytes);
    }
    catch (error) {
        return false;
    }
}
function sortObjectKeys(obj) {
    if (obj === null || typeof obj !== 'object') {
        return obj;
    }
    if (Array.isArray(obj)) {
        return obj.map(sortObjectKeys);
    }
    return Object.keys(obj)
        .sort()
        .reduce((result, key) => {
        result[key] = sortObjectKeys(obj[key]);
        return result;
    }, {});
}
function validateXPayment(payment) {
    const requiredShape = {
        x402Version: 'number',
        scheme: 'string',
        network: 'string',
        payload: {
            authorization: {
                from: 'string',
                to: 'string',
                value: 'string',
                validAfter: 'string',
                validBefore: 'string',
                nonce: 'string',
            },
            signature: 'string',
        },
    };
    const missing = checkShape(requiredShape, payment, '');
    if (missing.length > 0) {
        return missing.join('; ');
    }
    return 'valid';
}
function checkShape(expected, actual, path) {
    const missing = new Array();
    for (const key in expected) {
        const currentPath = path ? path + '.' + key : key;
        if (!(key in actual)) {
            missing.push('Missing field: ' + currentPath);
        }
        else if (typeof expected[key] === 'object') {
            if (typeof actual[key] !== 'object' || actual[key] === null) {
                missing.push('Invalid type at ' + currentPath + ': expected object');
            }
            else {
                checkShape(expected[key], actual[key], currentPath);
            }
        }
        else {
            if (typeof actual[key] !== expected[key]) {
                missing.push('Invalid type at ' +
                    currentPath +
                    ': expected ' +
                    expected[key] +
                    ', got ' +
                    typeof actual[key]);
            }
        }
    }
    return missing;
}
function verifyPaymentDetails(header, paymentRequirements) {
    const errors = [];
    const network = header.network;
    const configEntry = paymentRequirements.find((pc) => pc.network.toLowerCase() == (network || '').toLowerCase());
    if (configEntry == null) {
        errors.push('Invalid or unsupported network: ' + network);
    }
    if (configEntry) {
        try {
            const required = BigInt(configEntry.maxAmountRequired);
            let actual;
            actual = BigInt(header.payload.authorization.value);
            if (typeof actual !== 'undefined' && actual < required) {
                errors.push(`Value too low: got ${actual}, requires at least ${required}`);
            }
        }
        catch (e) {
            errors.push('Invalid value: must be numeric string');
        }
        const toAddr = header.payload?.authorization?.to;
        if (toAddr == null) {
            errors.push("Missing 'to' field in authorization");
        }
        else if (toAddr.toLowerCase() != configEntry.payTo.toLowerCase()) {
            errors.push(`Invalid 'to' address: expected ${configEntry.payTo}, got ${toAddr}`);
        }
        const now = Math.floor(Date.now() / 1000);
        try {
            const validAfter = Number(header.payload.authorization.validAfter);
            const validBefore = Number(header.payload.authorization.validBefore);
            if (validAfter > now) {
                errors.push(`Payment has not activated, validAfter is ${validAfter} but the server time is ${now}`);
            }
            if (validBefore < now) {
                errors.push(`Payment has expired, validBefore is ${validBefore} but the server time is ${now}`);
            }
        }
        catch (e) {
            errors.push(`Invalid validAfter or validBefore timestamps`);
        }
    }
    return {
        valid: errors.length == 0,
        errors: errors.join('; '),
        paymentRequirements: configEntry,
    };
}
async function registerWebhookWithX402Scan(context, webhookUrl) {
    context.logger.info(`Registering webhook with X402Scan: ${webhookUrl}`);
    const response = await context.helpers.request.call(context, {
        method: 'POST',
        url: 'https://www.x402scan.com/api/trpc/public.resources.register',
        body: {
            '0': {
                json: {
                    url: webhookUrl,
                    headers: {},
                },
            },
        },
        qs: {
            batch: 1,
        },
        headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
        },
        json: true,
        timeout: 15000,
    });
    context.logger.debug('X402Scan registration response', response);
}
class PaymentRequirements {
    constructor(scheme, network, maxAmountRequired, resource, description, mimeType, outputSchema, payTo, maxTimeoutSeconds, asset, extra) {
        this.scheme = scheme;
        this.network = network;
        this.maxAmountRequired = maxAmountRequired;
        this.resource = resource;
        this.description = description;
        this.mimeType = mimeType;
        this.outputSchema = outputSchema;
        this.payTo = payTo;
        this.maxTimeoutSeconds = maxTimeoutSeconds;
        this.asset = asset;
        this.extra = extra;
    }
}
