// n8n-nodes-1shot/dist/nodes/OneShot/executions/x402.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getX402Supported = getX402Supported;
exports.verifyX402Payment = verifyX402Payment;
exports.settleX402Payment = settleX402Payment;
const constants_1 = require("../types/constants");
class X402SupportedCacheEntry {
    constructor(timestamp, response) {
        this.timestamp = timestamp;
        this.response = response;
    }
}
const x402SupportedCache = new Map();
async function getX402Supported(context, ignoreCache = false) {
    try {
        const credentials = await context.getCredentials('oneShotOAuth2Api');
        const clientId = credentials.clientId;
        const cachedEntry = x402SupportedCache.get(clientId);
        if (cachedEntry && cachedEntry.timestamp > Date.now() - 1000 * 60 * 5 && !ignoreCache) {
            return cachedEntry.response;
        }
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'GET',
            url: '/x402/supported',
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        x402SupportedCache.set(clientId, new X402SupportedCacheEntry(Date.now(), response));
        return response;
    }
    catch (error) {
        context.logger.error(`Error getting X402 supported methods ${error.message}`, { error });
        throw error;
    }
}
async function verifyX402Payment(context, x402Version, paymentPayload, paymentRequirements) {
    try {
        const requestBody = {
            x402Version,
            paymentPayload,
            paymentRequirements,
        };
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: '/x402/verify',
            body: requestBody,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error verifying X402 payment ${error.message}`, { error });
        throw error;
    }
}
async function settleX402Payment(context, x402Version, paymentPayload, paymentRequirements) {
    try {
        const requestBody = {
            x402Version,
            paymentPayload,
            paymentRequirements,
        };
        const response = await context.helpers.requestWithAuthentication.call(context, 'oneShotOAuth2Api', {
            method: 'POST',
            url: '/x402/settle',
            body: requestBody,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            json: true,
            baseURL: constants_1.oneshotApiBaseUrl,
            timeout: 0,
        }, constants_1.additionalCredentialOptions);
        return response;
    }
    catch (error) {
        context.logger.error(`Error settling X402 payment ${error.message}`, { error });
        throw error;
    }
}
