// n8n-nodes-1shot/dist/credentials/oneShotOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.oneShotOAuth2Api = void 0;
const constants_1 = require("../nodes/OneShot/types/constants");
class oneShotOAuth2Api {
    constructor() {
        this.name = 'oneShotOAuth2Api';
        this.extends = ['oAuth2Api'];
        this.displayName = '1Shot OAuth2 API';
        this.documentationUrl = 'https://docs.1shotapi.com';
        this.properties = [
            {
                displayName: 'Grant Type',
                name: 'grantType',
                type: 'hidden',
                default: 'clientCredentials',
            },
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: '',
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'hidden',
                default: `${constants_1.oneshotApiBaseUrl}/token`,
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: '',
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: '',
            },
            {
                displayName: 'Authentication',
                name: 'authentication',
                type: 'hidden',
                default: 'body',
            },
            {
                displayName: 'Business ID',
                name: 'businessId',
                type: 'string',
                default: '',
            },
        ];
    }
}
exports.oneShotOAuth2Api = oneShotOAuth2Api;
