// n8n-nodes-1shot/dist/nodes/OneShot/descriptions/ChainDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chainOperationsFields = exports.chainOperations = void 0;
exports.chainOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['chains'],
            },
        },
        options: [
            {
                name: 'List Chains',
                value: 'list',
                description: 'List chains supported by 1Shot API',
                action: 'List chains',
            },
        ],
        default: 'list',
    },
];
const chainFields = [
    {
        displayName: 'Page Number',
        name: 'page',
        type: 'number',
        required: true,
        displayOptions: {
            show: {
                resource: ['chains'],
                operation: ['list'],
            },
        },
        default: 1,
        description: 'Enter the page number to get. This starts at 1.',
    },
    {
        displayName: 'Page Size',
        name: 'pageSize',
        type: 'number',
        required: true,
        displayOptions: {
            show: {
                resource: ['chains'],
                operation: ['list'],
            },
        },
        default: 25,
        description: 'Enter the size of the page to get',
    },
];
exports.chainOperationsFields = [...exports.chainOperations, ...chainFields];
