// n8n-nodes-1shot/dist/nodes/OneShot/descriptions/PromptDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.promptOperationsFields = exports.promptOperations = void 0;
const CommonDescriptions_1 = require("./CommonDescriptions");
exports.promptOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['prompts'],
            },
        },
        options: [
            {
                name: 'Search Prompts',
                value: 'search',
                description: 'Search for prompts with semantic search',
                action: 'Search prompts',
            },
        ],
        default: 'search',
    },
];
const promptFields = [
    {
        displayName: 'Query',
        name: 'query',
        type: 'string',
        default: '',
        description: 'Enter a query to search 1Shot Prompts for tools you can use in your dynamic workflow',
        required: true,
        displayOptions: {
            show: {
                resource: ['prompts'],
                operation: ['search'],
            },
        },
    },
    (0, CommonDescriptions_1.createChain)(false, 'prompts', ['search']),
];
exports.promptOperationsFields = [...exports.promptOperations, ...promptFields];
