// n8n-nodes-1shot/dist/nodes/OneShot/descriptions/X402RequestDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.x402RequestProperties = void 0;
exports.x402RequestProperties = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['x402Request'],
            },
        },
        options: [
            {
                name: 'X402 HTTP Request',
                value: 'x402',
                description: 'Make an HTTP request to an x402-enabled API endpoint. Any fee will be paid using a 1Shot Wallet.',
                action: 'X402 HTTP Request',
            },
        ],
        default: 'x402',
    },
    {
        displayName: '',
        name: 'curlImport',
        type: 'curlImport',
        default: '',
        displayOptions: {
            show: {
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'Wallet Name or ID',
        name: 'walletId',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'loadWalletOptions',
        },
        required: true,
        displayOptions: {
            show: {
                resource: ['x402Request'],
            },
        },
        default: '',
        description: 'Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code/expressions/">expression</a>',
    },
    {
        displayName: 'Method',
        name: 'method',
        type: 'options',
        options: [
            {
                name: 'DELETE',
                value: 'DELETE',
            },
            {
                name: 'GET',
                value: 'GET',
            },
            {
                name: 'HEAD',
                value: 'HEAD',
            },
            {
                name: 'OPTIONS',
                value: 'OPTIONS',
            },
            {
                name: 'PATCH',
                value: 'PATCH',
            },
            {
                name: 'POST',
                value: 'POST',
            },
            {
                name: 'PUT',
                value: 'PUT',
            },
        ],
        default: 'GET',
        description: 'The request method to use',
        displayOptions: {
            show: {
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'URL',
        name: 'url',
        type: 'string',
        default: '',
        placeholder: 'http://example.com/index.html',
        description: 'The URL to make the request to',
        required: true,
        displayOptions: {
            show: {
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'Authentication',
        name: 'authentication',
        noDataExpression: true,
        type: 'options',
        options: [
            {
                name: 'None',
                value: 'none',
            },
            {
                name: 'Predefined Credential Type',
                value: 'predefinedCredentialType',
                description: "We've already implemented auth for many services so that you don't have to set it up manually",
            },
            {
                name: 'Generic Credential Type',
                value: 'genericCredentialType',
                description: 'Fully customizable. Choose between basic, header, OAuth2, etc.',
            },
        ],
        default: 'none',
        displayOptions: {
            show: {
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'Credential Type',
        name: 'nodeCredentialType',
        type: 'credentialsSelect',
        noDataExpression: true,
        required: true,
        default: '',
        credentialTypes: ['extends:oAuth2Api', 'extends:oAuth1Api', 'has:authenticate'],
        displayOptions: {
            show: {
                resource: ['x402Request'],
                authentication: ['predefinedCredentialType'],
            },
        },
    },
    {
        displayName: 'Make sure you have specified the scope(s) for the Service Account in the credential',
        name: 'googleApiWarning',
        type: 'notice',
        default: '',
        displayOptions: {
            show: {
                nodeCredentialType: ['googleApi'],
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'Generic Auth Type',
        name: 'genericAuthType',
        type: 'credentialsSelect',
        required: true,
        default: '',
        credentialTypes: ['has:genericAuth'],
        displayOptions: {
            show: {
                authentication: ['genericCredentialType'],
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'SSL Certificates',
        name: 'provideSslCertificates',
        type: 'boolean',
        default: false,
        isNodeSetting: true,
        displayOptions: {
            show: {
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: "Provide certificates in node's 'Credential for SSL Certificates' parameter",
        name: 'provideSslCertificatesNotice',
        type: 'notice',
        default: '',
        isNodeSetting: true,
        displayOptions: {
            show: {
                provideSslCertificates: [true],
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'SSL Certificate',
        name: 'sslCertificate',
        type: 'credentials',
        default: '',
        displayOptions: {
            show: {
                provideSslCertificates: [true],
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'Send Query Parameters',
        name: 'sendQuery',
        type: 'boolean',
        default: false,
        noDataExpression: true,
        description: 'Whether the request has query params or not',
        displayOptions: {
            show: {
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'Specify Query Parameters',
        name: 'specifyQuery',
        type: 'options',
        displayOptions: {
            show: {
                sendQuery: [true],
                resource: ['x402Request'],
            },
        },
        options: [
            {
                name: 'Using Fields Below',
                value: 'keypair',
            },
            {
                name: 'Using JSON',
                value: 'json',
            },
        ],
        default: 'keypair',
    },
    {
        displayName: 'Query Parameters',
        name: 'queryParameters',
        type: 'fixedCollection',
        displayOptions: {
            show: {
                sendQuery: [true],
                specifyQuery: ['keypair'],
                resource: ['x402Request'],
            },
        },
        typeOptions: {
            multipleValues: true,
        },
        placeholder: 'Add Parameter',
        default: {
            parameters: [
                {
                    name: '',
                    value: '',
                },
            ],
        },
        options: [
            {
                name: 'parameters',
                displayName: 'Parameter',
                values: [
                    {
                        displayName: 'Name',
                        name: 'name',
                        type: 'string',
                        default: '',
                    },
                    {
                        displayName: 'Value',
                        name: 'value',
                        type: 'string',
                        default: '',
                    },
                ],
            },
        ],
    },
    {
        displayName: 'JSON',
        name: 'jsonQuery',
        type: 'json',
        displayOptions: {
            show: {
                sendQuery: [true],
                specifyQuery: ['json'],
                resource: ['x402Request'],
            },
        },
        default: '',
    },
    {
        displayName: 'Send Headers',
        name: 'sendHeaders',
        type: 'boolean',
        displayOptions: {
            show: {
                resource: ['x402Request'],
            },
        },
        default: false,
        noDataExpression: true,
        description: 'Whether the request has headers or not',
    },
    {
        displayName: 'Specify Headers',
        name: 'specifyHeaders',
        type: 'options',
        displayOptions: {
            show: {
                sendHeaders: [true],
                resource: ['x402Request'],
            },
        },
        options: [
            {
                name: 'Using Fields Below',
                value: 'keypair',
            },
            {
                name: 'Using JSON',
                value: 'json',
            },
        ],
        default: 'keypair',
    },
    {
        displayName: 'Header Parameters',
        name: 'headerParameters',
        type: 'fixedCollection',
        displayOptions: {
            show: {
                sendHeaders: [true],
                specifyHeaders: ['keypair'],
                resource: ['x402Request'],
            },
        },
        typeOptions: {
            multipleValues: true,
        },
        placeholder: 'Add Parameter',
        default: {
            parameters: [
                {
                    name: '',
                    value: '',
                },
            ],
        },
        options: [
            {
                name: 'parameters',
                displayName: 'Parameter',
                values: [
                    {
                        displayName: 'Name',
                        name: 'name',
                        type: 'string',
                        default: '',
                    },
                    {
                        displayName: 'Value',
                        name: 'value',
                        type: 'string',
                        default: '',
                    },
                ],
            },
        ],
    },
    {
        displayName: 'JSON',
        name: 'jsonHeaders',
        type: 'json',
        displayOptions: {
            show: {
                sendHeaders: [true],
                specifyHeaders: ['json'],
                resource: ['x402Request'],
            },
        },
        default: '',
    },
    {
        displayName: 'Send Body',
        name: 'sendBody',
        type: 'boolean',
        default: false,
        noDataExpression: true,
        description: 'Whether the request has a body or not',
        displayOptions: {
            show: {
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'Body Content Type',
        name: 'contentType',
        type: 'options',
        displayOptions: {
            show: {
                sendBody: [true],
                resource: ['x402Request'],
            },
        },
        options: [
            {
                name: 'Form Urlencoded',
                value: 'form-urlencoded',
            },
            {
                name: 'Form-Data',
                value: 'multipart-form-data',
            },
            {
                name: 'JSON',
                value: 'json',
            },
            {
                name: 'n8n Binary File',
                value: 'binaryData',
            },
            {
                name: 'Raw',
                value: 'raw',
            },
        ],
        default: 'json',
        description: 'Content-Type to use to send body parameters',
    },
    {
        displayName: 'Specify Body',
        name: 'specifyBody',
        type: 'options',
        displayOptions: {
            show: {
                sendBody: [true],
                contentType: ['json'],
                resource: ['x402Request'],
            },
        },
        options: [
            {
                name: 'Using Fields Below',
                value: 'keypair',
            },
            {
                name: 'Using JSON',
                value: 'json',
            },
        ],
        default: 'keypair',
        description: 'The body can be specified using explicit fields (<code>keypair</code>) or using a JavaScript object (<code>json</code>)',
    },
    {
        displayName: 'Body Parameters',
        name: 'bodyParameters',
        type: 'fixedCollection',
        displayOptions: {
            show: {
                sendBody: [true],
                contentType: ['json'],
                specifyBody: ['keypair'],
                resource: ['x402Request'],
            },
        },
        typeOptions: {
            multipleValues: true,
        },
        placeholder: 'Add Parameter',
        default: {
            parameters: [
                {
                    name: '',
                    value: '',
                },
            ],
        },
        options: [
            {
                name: 'parameters',
                displayName: 'Parameter',
                values: [
                    {
                        displayName: 'Name',
                        name: 'name',
                        type: 'string',
                        default: '',
                        description: 'ID of the field to set. Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code/expressions/">expression</a>.',
                    },
                    {
                        displayName: 'Value',
                        name: 'value',
                        type: 'string',
                        default: '',
                        description: 'Value of the field to set',
                    },
                ],
            },
        ],
    },
    {
        displayName: 'JSON',
        name: 'jsonBody',
        type: 'json',
        displayOptions: {
            show: {
                sendBody: [true],
                contentType: ['json'],
                specifyBody: ['json'],
                resource: ['x402Request'],
            },
        },
        default: '',
    },
    {
        displayName: 'Body Parameters',
        name: 'bodyParameters',
        type: 'fixedCollection',
        displayOptions: {
            show: {
                sendBody: [true],
                contentType: ['multipart-form-data'],
                resource: ['x402Request'],
            },
        },
        typeOptions: {
            multipleValues: true,
        },
        placeholder: 'Add Parameter',
        default: {
            parameters: [
                {
                    name: '',
                    value: '',
                },
            ],
        },
        options: [
            {
                name: 'parameters',
                displayName: 'Parameter',
                values: [
                    {
                        displayName: 'Parameter Type',
                        name: 'parameterType',
                        type: 'options',
                        options: [
                            {
                                name: 'n8n Binary File',
                                value: 'formBinaryData',
                            },
                            {
                                name: 'Form Data',
                                value: 'formData',
                            },
                        ],
                        default: 'formData',
                    },
                    {
                        displayName: 'Name',
                        name: 'name',
                        type: 'string',
                        default: '',
                        description: 'ID of the field to set. Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code/expressions/">expression</a>.',
                    },
                    {
                        displayName: 'Value',
                        name: 'value',
                        type: 'string',
                        displayOptions: {
                            show: {
                                parameterType: ['formData'],
                            },
                        },
                        default: '',
                        description: 'Value of the field to set',
                    },
                    {
                        displayName: 'Input Data Field Name',
                        name: 'inputDataFieldName',
                        type: 'string',
                        displayOptions: {
                            show: {
                                parameterType: ['formBinaryData'],
                            },
                        },
                        default: '',
                        description: 'The name of the incoming field containing the binary file data to be processed',
                    },
                ],
            },
        ],
    },
    {
        displayName: 'Specify Body',
        name: 'specifyBody',
        type: 'options',
        displayOptions: {
            show: {
                sendBody: [true],
                contentType: ['form-urlencoded'],
                resource: ['x402Request'],
            },
        },
        options: [
            {
                name: 'Using Fields Below',
                value: 'keypair',
            },
            {
                name: 'Using Single Field',
                value: 'string',
            },
        ],
        default: 'keypair',
    },
    {
        displayName: 'Body Parameters',
        name: 'bodyParameters',
        type: 'fixedCollection',
        displayOptions: {
            show: {
                sendBody: [true],
                contentType: ['form-urlencoded'],
                specifyBody: ['keypair'],
                resource: ['x402Request'],
            },
        },
        typeOptions: {
            multipleValues: true,
        },
        placeholder: 'Add Parameter',
        default: {
            parameters: [
                {
                    name: '',
                    value: '',
                },
            ],
        },
        options: [
            {
                name: 'parameters',
                displayName: 'Parameter',
                values: [
                    {
                        displayName: 'Name',
                        name: 'name',
                        type: 'string',
                        default: '',
                        description: 'ID of the field to set. Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code/expressions/">expression</a>.',
                    },
                    {
                        displayName: 'Value',
                        name: 'value',
                        type: 'string',
                        default: '',
                        description: 'Value of the field to set',
                    },
                ],
            },
        ],
    },
    {
        displayName: 'Body',
        name: 'body',
        type: 'string',
        displayOptions: {
            show: {
                sendBody: [true],
                specifyBody: ['string'],
                resource: ['x402Request'],
            },
        },
        default: '',
        placeholder: 'field1=value1&field2=value2',
    },
    {
        displayName: 'Input Data Field Name',
        name: 'inputDataFieldName',
        type: 'string',
        displayOptions: {
            show: {
                sendBody: [true],
                contentType: ['binaryData'],
                resource: ['x402Request'],
            },
        },
        default: '',
        description: 'The name of the incoming field containing the binary file data to be processed',
    },
    {
        displayName: 'Content Type',
        name: 'rawContentType',
        type: 'string',
        displayOptions: {
            show: {
                sendBody: [true],
                contentType: ['raw'],
                resource: ['x402Request'],
            },
        },
        default: '',
        placeholder: 'text/html',
    },
    {
        displayName: 'Body',
        name: 'body',
        type: 'string',
        displayOptions: {
            show: {
                sendBody: [true],
                contentType: ['raw'],
                resource: ['x402Request'],
            },
        },
        default: '',
        placeholder: '',
    },
    {
        displayName: 'Options',
        name: 'options',
        type: 'collection',
        placeholder: 'Add option',
        default: {},
        displayOptions: {
            show: {
                resource: ['x402Request'],
            },
        },
        options: [
            {
                displayName: 'Batching',
                name: 'batching',
                placeholder: 'Add Batching',
                type: 'fixedCollection',
                typeOptions: {
                    multipleValues: false,
                },
                default: {
                    batch: {},
                },
                options: [
                    {
                        displayName: 'Batching',
                        name: 'batch',
                        values: [
                            {
                                displayName: 'Items per Batch',
                                name: 'batchSize',
                                type: 'number',
                                typeOptions: {
                                    minValue: -1,
                                },
                                default: 50,
                                description: 'Input will be split in batches to throttle requests. -1 for disabled. 0 will be treated as 1.',
                            },
                            {
                                displayName: 'Batch Interval (ms)',
                                name: 'batchInterval',
                                type: 'number',
                                typeOptions: {
                                    minValue: 0,
                                },
                                default: 1000,
                                description: 'Time (in milliseconds) between each batch of requests. 0 for disabled.',
                            },
                        ],
                    },
                ],
            },
            {
                displayName: 'Ignore SSL Issues (Insecure)',
                name: 'allowUnauthorizedCerts',
                type: 'boolean',
                noDataExpression: true,
                default: false,
                description: 'Whether to download the response even if SSL certificate validation is not possible',
            },
            {
                displayName: 'Array Format in Query Parameters',
                name: 'queryParameterArrays',
                type: 'options',
                displayOptions: {
                    show: {
                        '/sendQuery': [true],
                    },
                },
                options: [
                    {
                        name: 'No Brackets',
                        value: 'repeat',
                        description: 'e.g. foo=bar&foo=qux',
                    },
                    {
                        name: 'Brackets Only',
                        value: 'brackets',
                        description: 'e.g. foo[]=bar&foo[]=qux',
                    },
                    {
                        name: 'Brackets with Indices',
                        value: 'indices',
                        description: 'e.g. foo[0]=bar&foo[1]=qux',
                    },
                ],
                default: 'brackets',
            },
            {
                displayName: 'Lowercase Headers',
                name: 'lowercaseHeaders',
                type: 'boolean',
                default: true,
                description: 'Whether to lowercase header names',
            },
            {
                displayName: 'Redirects',
                name: 'redirect',
                placeholder: 'Add Redirect',
                type: 'fixedCollection',
                typeOptions: {
                    multipleValues: false,
                },
                default: { redirect: {} },
                options: [
                    {
                        displayName: 'Redirect',
                        name: 'redirect',
                        values: [
                            {
                                displayName: 'Follow Redirects',
                                name: 'followRedirects',
                                type: 'boolean',
                                default: false,
                                noDataExpression: true,
                                description: 'Whether to follow all redirects',
                            },
                            {
                                displayName: 'Max Redirects',
                                name: 'maxRedirects',
                                type: 'number',
                                displayOptions: {
                                    show: {
                                        followRedirects: [true],
                                    },
                                },
                                default: 21,
                                description: 'Max number of redirects to follow',
                            },
                        ],
                    },
                ],
                displayOptions: {
                    show: {
                        '@version': [1, 2, 3],
                    },
                },
            },
            {
                displayName: 'Redirects',
                name: 'redirect',
                placeholder: 'Add Redirect',
                type: 'fixedCollection',
                typeOptions: {
                    multipleValues: false,
                },
                default: {
                    redirect: {},
                },
                options: [
                    {
                        displayName: 'Redirect',
                        name: 'redirect',
                        values: [
                            {
                                displayName: 'Follow Redirects',
                                name: 'followRedirects',
                                type: 'boolean',
                                default: true,
                                noDataExpression: true,
                                description: 'Whether to follow all redirects',
                            },
                            {
                                displayName: 'Max Redirects',
                                name: 'maxRedirects',
                                type: 'number',
                                displayOptions: {
                                    show: {
                                        followRedirects: [true],
                                    },
                                },
                                default: 21,
                                description: 'Max number of redirects to follow',
                            },
                        ],
                    },
                ],
                displayOptions: {},
            },
            {
                displayName: 'Response',
                name: 'response',
                placeholder: 'Add response',
                type: 'fixedCollection',
                typeOptions: {
                    multipleValues: false,
                },
                default: {
                    response: {},
                },
                options: [
                    {
                        displayName: 'Response',
                        name: 'response',
                        values: [
                            {
                                displayName: 'Include Response Headers and Status',
                                name: 'fullResponse',
                                type: 'boolean',
                                default: false,
                                description: 'Whether to return the full response (headers and response status code) data instead of only the body',
                            },
                            {
                                displayName: 'Never Error',
                                name: 'neverError',
                                type: 'boolean',
                                default: false,
                                description: 'Whether to succeeds also when status code is not 2xx',
                            },
                            {
                                displayName: 'Response Format',
                                name: 'responseFormat',
                                type: 'options',
                                noDataExpression: true,
                                options: [
                                    {
                                        name: 'Autodetect',
                                        value: 'autodetect',
                                    },
                                    {
                                        name: 'File',
                                        value: 'file',
                                    },
                                    {
                                        name: 'JSON',
                                        value: 'json',
                                    },
                                    {
                                        name: 'Text',
                                        value: 'text',
                                    },
                                ],
                                default: 'autodetect',
                                description: 'The format in which the data gets returned from the URL',
                            },
                            {
                                displayName: 'Put Output in Field',
                                name: 'outputPropertyName',
                                type: 'string',
                                default: 'data',
                                required: true,
                                displayOptions: {
                                    show: {
                                        responseFormat: ['file', 'text'],
                                    },
                                },
                                description: 'Name of the binary property to which to write the data of the read file',
                            },
                        ],
                    },
                ],
            },
            {
                displayName: 'Timeout',
                name: 'timeout',
                type: 'number',
                typeOptions: {
                    minValue: 1,
                },
                default: 10000,
                description: 'Time in ms to wait for the server to send response headers (and start the response body) before aborting the request',
            },
        ],
    },
    {
        displayName: 'Optimize Response',
        name: 'optimizeResponse',
        type: 'boolean',
        default: false,
        noDataExpression: true,
        description: 'Whether the optimize the tool response to reduce amount of data passed to the LLM that could lead to better result and reduce cost',
        displayOptions: {
            show: {
                '@tool': [true],
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'Expected Response Type',
        name: 'responseType',
        type: 'options',
        displayOptions: {
            show: {
                optimizeResponse: [true],
                '@tool': [true],
                resource: ['x402Request'],
            },
        },
        options: [
            {
                name: 'JSON',
                value: 'json',
            },
            {
                name: 'HTML',
                value: 'html',
            },
            {
                name: 'Text',
                value: 'text',
            },
        ],
        default: 'json',
    },
    {
        displayName: 'Field Containing Data',
        name: 'dataField',
        type: 'string',
        default: '',
        placeholder: 'e.g. records',
        description: 'Specify the name of the field in the response containing the data',
        hint: 'leave blank to use whole response',
        requiresDataPath: 'single',
        displayOptions: {
            show: {
                optimizeResponse: [true],
                responseType: ['json'],
                '@tool': [true],
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'Include Fields',
        name: 'fieldsToInclude',
        type: 'options',
        description: 'What fields response object should include',
        default: 'all',
        displayOptions: {
            show: {
                optimizeResponse: [true],
                responseType: ['json'],
                '@tool': [true],
                resource: ['x402Request'],
            },
        },
        options: [
            {
                name: 'All',
                value: 'all',
                description: 'Include all fields',
            },
            {
                name: 'Selected',
                value: 'selected',
                description: 'Include only fields specified below',
            },
            {
                name: 'Except',
                value: 'except',
                description: 'Exclude fields specified below',
            },
        ],
    },
    {
        displayName: 'Fields',
        name: 'fields',
        type: 'string',
        default: '',
        placeholder: 'e.g. field1,field2',
        description: 'Comma-separated list of the field names. Supports dot notation. You can drag the selected fields from the input panel.',
        requiresDataPath: 'multiple',
        displayOptions: {
            show: {
                optimizeResponse: [true],
                responseType: ['json'],
                '@tool': [true],
                resource: ['x402Request'],
            },
            hide: {
                fieldsToInclude: ['all'],
            },
        },
    },
    {
        displayName: 'Selector (CSS)',
        name: 'cssSelector',
        type: 'string',
        description: 'Select specific element(e.g. body) or multiple elements(e.g. div) of chosen type in the response HTML.',
        placeholder: 'e.g. body',
        default: 'body',
        displayOptions: {
            show: {
                optimizeResponse: [true],
                responseType: ['html'],
                '@tool': [true],
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'Return Only Content',
        name: 'onlyContent',
        type: 'boolean',
        default: false,
        description: 'Whether to return only content of html elements, stripping html tags and attributes',
        hint: 'Uses less tokens and may be easier for model to understand',
        displayOptions: {
            show: {
                optimizeResponse: [true],
                responseType: ['html'],
                '@tool': [true],
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'Elements To Omit',
        name: 'elementsToOmit',
        type: 'string',
        displayOptions: {
            show: {
                optimizeResponse: [true],
                responseType: ['html'],
                onlyContent: [true],
                '@tool': [true],
                resource: ['x402Request'],
            },
        },
        default: '',
        placeholder: 'e.g. img, .className, #ItemId',
        description: 'Comma-separated list of selectors that would be excluded when extracting content',
    },
    {
        displayName: 'Truncate Response',
        name: 'truncateResponse',
        type: 'boolean',
        default: false,
        hint: 'Helps save tokens',
        displayOptions: {
            show: {
                optimizeResponse: [true],
                responseType: ['text', 'html'],
                '@tool': [true],
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: 'Max Response Characters',
        name: 'maxLength',
        type: 'number',
        default: 1000,
        typeOptions: {
            minValue: 1,
        },
        displayOptions: {
            show: {
                optimizeResponse: [true],
                responseType: ['text', 'html'],
                truncateResponse: [true],
                '@tool': [true],
                resource: ['x402Request'],
            },
        },
    },
    {
        displayName: "You can view the raw requests this node makes in your browser's developer console",
        name: 'infoMessage',
        type: 'notice',
        default: '',
        displayOptions: {
            show: {
                resource: ['x402Request'],
            },
        },
    },
];
