// n8n-nodes-1shot/dist/nodes/OneShot/utils/genericRequestFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updadeQueryParameterConfig = exports.setAgentOptions = exports.prepareRequestBody = exports.binaryContentTypes = exports.getOAuth2AdditionalParameters = exports.REDACTED = exports.replaceNullValues = void 0;
exports.sanitizeUiMessage = sanitizeUiMessage;
exports.getSecrets = getSecrets;
exports.reduceAsync = reduceAsync;
exports.toText = toText;
exports.formatPrivateKey = formatPrivateKey;
const n8n_workflow_1 = require("n8n-workflow");
const n8nUtils_1 = require("./n8nUtils");
const lodashFunctions_1 = require("./lodashFunctions");
const replaceNullValues = (item) => {
    if (item.json === null) {
        item.json = {};
    }
    return item;
};
exports.replaceNullValues = replaceNullValues;
exports.REDACTED = '**hidden**';
function isObject(obj) {
    return !!obj && (obj.constructor === Object || Object.getPrototypeOf(obj) === null);
}
function redact(obj, secrets) {
    if (typeof obj === 'string') {
        return secrets.reduce((safe, secret) => safe.replace(secret, exports.REDACTED), obj);
    }
    if (Array.isArray(obj)) {
        return obj.map((item) => redact(item, secrets));
    }
    else if (isObject(obj)) {
        for (const [key, value] of Object.entries(obj)) {
            (0, n8nUtils_1.setSafeObjectProperty)(obj, key, redact(value, secrets));
        }
    }
    return obj;
}
function sanitizeUiMessage(request, authDataKeys, secrets) {
    const { body, ...rest } = request;
    let sendRequest = { body };
    for (const [key, value] of Object.entries(rest)) {
        sendRequest[key] = (0, n8n_workflow_1.deepCopy)(value);
    }
    if (Buffer.isBuffer(sendRequest.body) && sendRequest.body.length > 250000) {
        sendRequest = {
            ...request,
            body: `Binary data got replaced with this text. Original was a Buffer with a size of ${request.body.length} bytes.`,
        };
    }
    for (const requestProperty of Object.keys(authDataKeys)) {
        sendRequest = {
            ...sendRequest,
            [requestProperty]: Object.keys(sendRequest[requestProperty]).reduce((acc, curr) => {
                acc[curr] = authDataKeys[requestProperty].includes(curr)
                    ? exports.REDACTED
                    : sendRequest[requestProperty][curr];
                return acc;
            }, {}),
        };
    }
    const HEADER_BLOCKLIST = new Set([
        'authorization',
        'x-api-key',
        'x-auth-token',
        'cookie',
        'proxy-authorization',
        'sslclientcert',
    ]);
    const headers = sendRequest.headers;
    if (headers) {
        for (const headerName of Object.keys(headers)) {
            if (HEADER_BLOCKLIST.has(headerName.toLowerCase())) {
                headers[headerName] = exports.REDACTED;
            }
        }
    }
    if (secrets && secrets.length > 0) {
        return redact(sendRequest, secrets);
    }
    return sendRequest;
}
function getSecrets(properties, credentials) {
    const sensitivePropNames = new Set(properties.filter((prop) => prop.typeOptions?.password).map((prop) => prop.name));
    const secrets = Object.entries(credentials)
        .filter(([propName]) => sensitivePropNames.has(propName))
        .map(([_, value]) => value)
        .filter((value) => typeof value === 'string');
    const oauthAccessToken = (0, lodashFunctions_1.getNestedProperty)(credentials, 'oauthTokenData.access_token');
    if (typeof oauthAccessToken === 'string') {
        secrets.push(oauthAccessToken);
    }
    return secrets;
}
const getOAuth2AdditionalParameters = (nodeCredentialType) => {
    const oAuth2Options = {
        bitlyOAuth2Api: {
            tokenType: 'Bearer',
        },
        boxOAuth2Api: {
            includeCredentialsOnRefreshOnBody: true,
        },
        ciscoWebexOAuth2Api: {
            tokenType: 'Bearer',
        },
        clickUpOAuth2Api: {
            keepBearer: false,
            tokenType: 'Bearer',
        },
        goToWebinarOAuth2Api: {
            tokenExpiredStatusCode: 403,
        },
        hubspotDeveloperApi: {
            tokenType: 'Bearer',
            includeCredentialsOnRefreshOnBody: true,
        },
        hubspotOAuth2Api: {
            tokenType: 'Bearer',
            includeCredentialsOnRefreshOnBody: true,
        },
        lineNotifyOAuth2Api: {
            tokenType: 'Bearer',
        },
        linkedInOAuth2Api: {
            tokenType: 'Bearer',
        },
        mailchimpOAuth2Api: {
            tokenType: 'Bearer',
        },
        mauticOAuth2Api: {
            includeCredentialsOnRefreshOnBody: true,
        },
        microsoftAzureMonitorOAuth2Api: {
            tokenExpiredStatusCode: 403,
        },
        microsoftDynamicsOAuth2Api: {
            property: 'id_token',
        },
        philipsHueOAuth2Api: {
            tokenType: 'Bearer',
        },
        raindropOAuth2Api: {
            includeCredentialsOnRefreshOnBody: true,
        },
        shopifyOAuth2Api: {
            tokenType: 'Bearer',
            keyToIncludeInAccessTokenHeader: 'X-Shopify-Access-Token',
        },
        slackOAuth2Api: {
            tokenType: 'Bearer',
            property: 'authed_user.access_token',
        },
        stravaOAuth2Api: {
            includeCredentialsOnRefreshOnBody: true,
        },
    };
    return oAuth2Options[nodeCredentialType];
};
exports.getOAuth2AdditionalParameters = getOAuth2AdditionalParameters;
exports.binaryContentTypes = [
    'image/',
    'audio/',
    'video/',
    'application/octet-stream',
    'application/gzip',
    'application/zip',
    'application/vnd.rar',
    'application/epub+zip',
    'application/x-bzip',
    'application/x-bzip2',
    'application/x-cdf',
    'application/vnd.amazon.ebook',
    'application/msword',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'application/vnd.ms-fontobject',
    'application/vnd.oasis.opendocument.presentation',
    'application/pdf',
    'application/x-tar',
    'application/vnd.visio',
    'application/vnd.ms-excel',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    'application/x-7z-compressed',
];
async function reduceAsync(arr, reducer, init = Promise.resolve({})) {
    return await arr.reduce(async (promiseAcc, item) => {
        return await reducer(await promiseAcc, item);
    }, init);
}
const prepareRequestBody = async (parameters, bodyType, version, defaultReducer) => {
    if (bodyType === 'json' && version >= 4) {
        return await parameters.reduce(async (acc, entry) => {
            const result = await acc;
            (0, lodashFunctions_1.setNestedProperty)(result, entry.name, entry.value);
            return result;
        }, Promise.resolve({}));
    }
    else if (bodyType === 'multipart-form-data' && version >= 4.2) {
        const formData = {};
        for (const parameter of parameters) {
            if (parameter.parameterType === 'formBinaryData') {
                const entry = await defaultReducer({}, parameter);
                const key = Object.keys(entry)[0];
                const data = entry[key];
                formData[key] = data;
                continue;
            }
            formData[parameter.name] = parameter.value;
        }
        return formData;
    }
    else {
        return await reduceAsync(parameters, defaultReducer);
    }
};
exports.prepareRequestBody = prepareRequestBody;
const setAgentOptions = (requestOptions, sslCertificates) => {
    if (sslCertificates) {
        const agentOptions = {};
        if (sslCertificates.ca)
            agentOptions.ca = formatPrivateKey(sslCertificates.ca);
        if (sslCertificates.cert)
            agentOptions.cert = formatPrivateKey(sslCertificates.cert);
        if (sslCertificates.key)
            agentOptions.key = formatPrivateKey(sslCertificates.key);
        if (sslCertificates.passphrase)
            agentOptions.passphrase = formatPrivateKey(sslCertificates.passphrase);
        requestOptions.agentOptions = agentOptions;
    }
};
exports.setAgentOptions = setAgentOptions;
const updadeQueryParameterConfig = (version) => {
    if (version < 4.3) {
        return (qs, name, value) => (qs[name] = value);
    }
    else {
        return (qs, name, value) => {
            if (qs[name] === undefined) {
                qs[name] = value;
            }
            else if (Array.isArray(qs[name])) {
                qs[name].push(value);
            }
            else {
                qs[name] = [qs[name], value];
            }
        };
    }
};
exports.updadeQueryParameterConfig = updadeQueryParameterConfig;
function toText(data) {
    if (typeof data === 'object' && data !== null) {
        return JSON.stringify(data);
    }
    return data;
}
function formatPrivateKey(privateKey, keyIsPublic = false) {
    let regex = /(PRIVATE KEY|CERTIFICATE)/;
    if (keyIsPublic) {
        regex = /(PUBLIC KEY)/;
    }
    if (!privateKey || /\n/.test(privateKey)) {
        return privateKey;
    }
    let formattedPrivateKey = '';
    const parts = privateKey.split('-----').filter((item) => item !== '');
    parts.forEach((part) => {
        if (regex.test(part)) {
            formattedPrivateKey += `-----${part}-----`;
        }
        else {
            const passRegex = /Proc-Type|DEK-Info/;
            if (passRegex.test(part)) {
                part = part.replace(/:\s+/g, ':');
                formattedPrivateKey += part.replace(/\\n/g, '\n').replace(/\s+/g, '\n');
            }
            else {
                formattedPrivateKey += part.replace(/\\n/g, '\n').replace(/\s+/g, '\n');
            }
        }
    });
    return formattedPrivateKey;
}
