// n8n-nodes-1shot/dist/nodes/OneShot/utils/webhookUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isIpWhitelisted = exports.setupOutputConnection = exports.configuredOutputs = exports.getResponseHeaders = exports.getResponseData = exports.getResponseCode = void 0;
exports.getX402RefundHeader = getX402RefundHeader;
const getResponseCode = (parameters) => {
    if (parameters.responseCode) {
        return parameters.responseCode;
    }
    const responseCodeOptions = parameters.options;
    if (responseCodeOptions?.responseCode?.values) {
        const { responseCode, customCode } = responseCodeOptions.responseCode.values;
        if (customCode) {
            return customCode;
        }
        return responseCode;
    }
    return 200;
};
exports.getResponseCode = getResponseCode;
const getResponseData = (parameters) => {
    const { responseData, responseMode, options } = parameters;
    if (responseData)
        return responseData;
    if (responseMode === 'onReceived') {
        const data = options?.responseData;
        if (data)
            return data;
    }
    if (options?.noResponseBody)
        return 'noData';
    return undefined;
};
exports.getResponseData = getResponseData;
const getResponseHeaders = (parameters) => {
    if (parameters.options == null) {
        return null;
    }
    let responseHeaders = parameters.options.responseHeaders;
    try {
        const x402RefundsContactEmail = parameters.options.x402RefundsContactEmail;
        if (x402RefundsContactEmail == null || x402RefundsContactEmail === '') {
            return null;
        }
        const refundContact = `<mailto:${x402RefundsContactEmail}>; rel="https://x402refunds.com/rel/refund-contact"`;
        const refundRequest = '<https://api.x402refunds.com/v1/refunds>; rel="https://x402refunds.com/rel/refund-request"; type="application/json"';
        const x402RefundsHeader = `${refundContact}, ${refundRequest}`;
        if (responseHeaders == null) {
            responseHeaders = {
                entries: [],
            };
        }
        const entries = responseHeaders.entries;
        const existingLinkIndex = entries.findIndex((entry) => entry.name?.toLowerCase() === 'link');
        if (existingLinkIndex >= 0) {
            const existing = entries[existingLinkIndex];
            existing.value = existing.value
                ? `${existing.value}, ${x402RefundsHeader}`
                : x402RefundsHeader;
            return responseHeaders;
        }
        entries.push({
            name: 'Link',
            value: x402RefundsHeader,
        });
        return responseHeaders;
    }
    catch (e) {
        return responseHeaders;
    }
};
exports.getResponseHeaders = getResponseHeaders;
function getX402RefundHeader(x402RefundsContactEmail) {
    if (x402RefundsContactEmail == null || x402RefundsContactEmail === '') {
        return null;
    }
    const refundContact = `<mailto:${x402RefundsContactEmail}>; rel="https://x402refunds.com/rel/refund-contact"`;
    const refundRequest = '<https://api.x402refunds.com/v1/refunds>; rel="https://x402refunds.com/rel/refund-request"; type="application/json"';
    return `${refundContact}, ${refundRequest}`;
}
const configuredOutputs = (parameters) => {
    const httpMethod = parameters.httpMethod;
    if (!Array.isArray(httpMethod))
        return [
            {
                type: 'main',
                displayName: httpMethod,
            },
        ];
    const outputs = httpMethod.map((method) => {
        return {
            type: 'main',
            displayName: method,
        };
    });
    return outputs;
};
exports.configuredOutputs = configuredOutputs;
const setupOutputConnection = (ctx, method, additionalData) => {
    const httpMethod = ctx.getNodeParameter('httpMethod', []);
    let webhookUrl = ctx.getNodeWebhookUrl('default');
    const executionMode = ctx.getMode() === 'manual' ? 'test' : 'production';
    if (executionMode === 'test') {
        webhookUrl = webhookUrl.replace('/webhook/', '/webhook-test/');
    }
    if (!Array.isArray(httpMethod)) {
        return (outputData) => {
            outputData.json.webhookUrl = webhookUrl;
            outputData.json.executionMode = executionMode;
            if (additionalData?.jwtPayload) {
                outputData.json.jwtPayload = additionalData.jwtPayload;
            }
            return [[outputData]];
        };
    }
    const outputIndex = httpMethod.indexOf(method.toUpperCase());
    const outputs = httpMethod.map(() => []);
    return (outputData) => {
        outputData.json.webhookUrl = webhookUrl;
        outputData.json.executionMode = executionMode;
        if (additionalData?.jwtPayload) {
            outputData.json.jwtPayload = additionalData.jwtPayload;
        }
        outputs[outputIndex] = [outputData];
        return outputs;
    };
};
exports.setupOutputConnection = setupOutputConnection;
const isIpWhitelisted = (whitelist, ips, ip) => {
    if (whitelist === undefined || whitelist === '') {
        return true;
    }
    if (!Array.isArray(whitelist)) {
        whitelist = whitelist.split(',').map((entry) => entry.trim());
    }
    for (const address of whitelist) {
        if (ip?.includes(address)) {
            return true;
        }
        if (ips.some((entry) => entry.includes(address))) {
            return true;
        }
    }
    return false;
};
exports.isIpWhitelisted = isIpWhitelisted;
